---
title: "Running wavess in Python"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Running wavess in Python}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

We provide the R function `run_wavess()` as a convenience function to run the
simulator from R, but since the underlying functions are all written in Python,
you can also run wavess with Python using a command-line script. We provide an
example Python script `main.py` (in `inst/python/`) to make this easy to do. The
script takes a config file (`inst/python/config.yaml`) that includes paths to
input files and defines various input arguments such as what selective pressures
to model, as well as an output file path/prefix.

We provide example input files in **XXXXXXX**. You can generate the input files
on your own, or using the functions provided in this package. See
`vignette("prepare_input_data")` for details on how to create and save the
required input files. Here, we briefly outline what goes in the config file and
how to run `main.py` itself.

### Installing dependencies

The Python script implementation of wavess requires the following imports:
pandas, Bio, time, sys, os, yaml, csv, copy, random, numpy, collections

The easiest way to install these is to create a conda environment.
f you don't already have it installed, first install 
[mamba](https://mamba.readthedocs.io/en/latest/installation/mamba-installation.html) 
(or conda, but it's slower). 

Then, create the wavess conda environment (you only have to do this once),
you can use the environment yaml file in the `inst/python` directory. Navigate
to that directory and then run the command:

```
mamba env create -f env.yaml
```

Next, activate the wavess environment (you have to do this once per session):

```
mamba activate swish
```

### Config file

`main.py` requires a `config.yaml` file (example config file in `inst/python`)
that defines various parameters and includes paths to input files.

**File inputs (example data in `inst/extdata/`):**

- `pop_samp`: tsv with columns generation, active_cell_count, n_sample_active.
Note that the initial active cell population size in the first generation must
be the same as the number of input founder sequences (because it simply _is_ the
input founder sequences). Can be generated using the [generate_pop_samp()]
function.
- `founder_seqs`: fasta file of founder sequence(s). The founder sequence(s) may
only contain the characters ACGT, and no gaps are allowed. When modeling immune
fitness, they are expected to be codon-aligned.
- `q`: csv that is a named Q matrix of nucleotide substitution rates. 
Rows are from, columns are to. Can be generated using the [estimate_q()]
function.
- `conserved_sites`: tsv of conserved sites indexed at 0 where each row contains
a conserved site. This can be generated using the [identify_conserved_sites()]
function. If this is an empty string, then conserved fitness costs are not included.
- `ref_seq`: fasta file of the reference sequence. A consensus
sequence, that can be used as the reference sequence, can be generated using the
function [identify_conserved_sites()]. If this is an empty string, then fitness
costs relative to a reference sequence are not included. 
- `epitope_locations`: tsv of epitope locations and maximum fitness
costs with columns epi_start_nt, epi_end_nt, max_fitness_cost. These epitopes
are expected to be indexed at 0 and in a protein in the correct reading frame,
as the nucleotide sequences are translated to amino acids to calculate the
immune fitness cost. This information can be generated using the functions
[get_epitope_frequencies()] and [sample_epitopes()]. If this is an empty string
provided, then immune fitness costs are not included.

**Parameters:**

Diversity-generating mechanisms:

- `mut_rate`: Mutation rate per site per generation
- `recomb_rate``: Recombination rate per site per generation

Latent cell dynamics:

- `prob_act_to_lat`: probability of infected cell becoming latent
- `prob_lat_to_act`: probability that a latent cell will become active
- `prob_lat_prolif`: probability that a latent cell will proliferate (based on central memory cells)
- `prob_lat_die`: probability that a latent cell will die

Parameters related to fitness:

Conserved sites:

- `conserved_cost`: fitness cost per mutation in a conserved site

Fitness relative to a reference sequence:

- `rep_exp`: exponent for replicative fitness function

Immune fitness:

- `seroconversion_time`: seroconversion time in generations; immunity considered only at this generation onward
- `immune_response_proportion`: proportion new viral antigen must reach to initiate immune response against corresponding epitope
- `time_to_full_potency`: number of generations for a new immune response to reach full potency

### Running `run_wavess.py`

`run_wavess.py` takes two required inputs and one optional input:

- The config file, which contains information about input file paths and parameters
- The prefix to the output files (including directories)
- An optional number that is the seed to set for random number generation (to make your results reproducible)

Here is an example (we set a seed so it will be reproducible, usually you don't want to do this):

```
python run_wavess.py config.yaml wavess_output/ 1234
```

**TODO THROW ERROR IF NO SLASH IN OUTPUT FILE**

**Note that the path to the output file has to have at least one slash.**
If the slash is at the end, then there will be no prefix for the output files.
If you want a prefix for the output files, then you can add that to the end.

Two output files are generated: a csv file containing various counts for each
sampled generation, and a fasta file containing the sampled sequences for each
generation. See `vignette("run_wavess")` for more details about the outputs. 

Example output files can be found in `inst/python/wavess_output/`.


### Python tests

To run the tests for the Python part of the code, navigate to `inst/python` and 
run:

```
pytest tests.py --cov --cov-report=html
```

You can then look at the coverage report in `htmlcov/index.html`.


