<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Run wavess — run_wavess • wavess</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Run wavess — run_wavess"><meta name="description" content="Simulate within-host evolution optionally including recombination (default:
on), latency (default: on), and fitness costs (default: off). The three
fitness costs that can be simulated are conserved sites, fitness relative to
a reference sequence, and antibody-based immune fitness costs. Nucleotide
positions for conserved and immune fitness are expected to be indexed at 0.
Please note that the default arguments were set with the the HIV env gp120
gene in mind. If you'd like to simulate something else, you will likely have
to modify certain parameters. However, if you are interested in this gene in
particular, you can probably use most of the defaults including the founder
and reference sequences provided as examples. However, by default no fitness
costs are modeled. We recommend including fitness to obtain a realistic model
output. To model these, you can use the pre-processing functions (see
vignette('prepare_input_data')) to generate the relevant inputs. Also, the
parameters for latent probabilities are assumed to be small, such that it is
unlikely that multiple events (activate, die, proliferate) will occur to a
single latent cell in a single (active cell) generation. See
vignette('run_wavess') for more details about the simulator and input
arguments."><meta property="og:description" content="Simulate within-host evolution optionally including recombination (default:
on), latency (default: on), and fitness costs (default: off). The three
fitness costs that can be simulated are conserved sites, fitness relative to
a reference sequence, and antibody-based immune fitness costs. Nucleotide
positions for conserved and immune fitness are expected to be indexed at 0.
Please note that the default arguments were set with the the HIV env gp120
gene in mind. If you'd like to simulate something else, you will likely have
to modify certain parameters. However, if you are interested in this gene in
particular, you can probably use most of the defaults including the founder
and reference sequences provided as examples. However, by default no fitness
costs are modeled. We recommend including fitness to obtain a realistic model
output. To model these, you can use the pre-processing functions (see
vignette('prepare_input_data')) to generate the relevant inputs. Also, the
parameters for latent probabilities are assumed to be small, such that it is
unlikely that multiple events (activate, die, proliferate) will occur to a
single latent cell in a single (active cell) generation. See
vignette('run_wavess') for more details about the simulator and input
arguments."><meta property="og:image" content="https://molevolepid.github.io/wavess/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">wavess</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/prepare_input_data.html">Prepare input data</a></li>
    <li><a class="dropdown-item" href="../articles/run_wavess.html">Simulate within-host evolution with `wavess`</a></li>
    <li><a class="dropdown-item" href="../articles/analyze_output.html">Analyze wavess output</a></li>
    <li><a class="dropdown-item" href="../articles/python.html">Running wavess in Python</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Run wavess</h1>

      <div class="d-none name"><code>run_wavess.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Simulate within-host evolution optionally including recombination (default:
on), latency (default: on), and fitness costs (default: off). The three
fitness costs that can be simulated are conserved sites, fitness relative to
a reference sequence, and antibody-based immune fitness costs. Nucleotide
positions for conserved and immune fitness are expected to be indexed at 0.
Please note that the default arguments were set with the the HIV <em>env</em> gp120
gene in mind. If you'd like to simulate something else, you will likely have
to modify certain parameters. However, if you are interested in this gene in
particular, you can probably use most of the defaults including the founder
and reference sequences provided as examples. However, by default no fitness
costs are modeled. We recommend including fitness to obtain a realistic model
output. To model these, you can use the pre-processing functions (see
<code><a href="../articles/prepare_input_data.html">vignette('prepare_input_data')</a></code>) to generate the relevant inputs. Also, the
parameters for latent probabilities are assumed to be small, such that it is
unlikely that multiple events (activate, die, proliferate) will occur to a
single latent cell in a single (active cell) generation. See
<code><a href="../articles/run_wavess.html">vignette('run_wavess')</a></code> for more details about the simulator and input
arguments.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_wavess</span><span class="op">(</span></span>
<span>  <span class="va">inf_pop_size</span>,</span>
<span>  <span class="va">samp_scheme</span>,</span>
<span>  <span class="va">founder_seqs</span>,</span>
<span>  generation_time <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  mut_rate <span class="op">=</span> <span class="fl">3e-05</span>,</span>
<span>  q <span class="op">=</span> <span class="fu">wavess</span><span class="fu">::</span><span class="fu"><a href="calc_q_from_rates.html">calc_q_from_rates</a></span><span class="op">(</span><span class="fu">wavess</span><span class="fu">::</span><span class="va"><a href="hiv_mut_rates.html">hiv_mut_rates</a></span>, <span class="va">mut_rate</span>, <span class="va">generation_time</span><span class="op">)</span>,</span>
<span>  recomb_rate <span class="op">=</span> <span class="fl">1.4e-05</span>,</span>
<span>  act_to_lat <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  lat_to_act <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  lat_prolif <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  lat_die <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  conserved_sites <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  conserved_cost <span class="op">=</span> <span class="fl">0.99</span>,</span>
<span>  ref_seq <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  replicative_cost <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  epitope_locations <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n_for_imm <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  days_full_potency <span class="op">=</span> <span class="fl">90</span>,</span>
<span>  immune_start_day <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  seed <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-inf-pop-size">inf_pop_size<a class="anchor" aria-label="anchor" href="#arg-inf-pop-size"></a></dt>
<dd><p>Tibble with columns <em>generation</em> (starting at day 0) and
active_cell_count. Note that the initial active cell population size on day
0 must be the same as the number of input founder sequences (because it
simply <em>is</em> the input founder sequences).  Can be generated using the
<code><a href="define_growth_curve.html">define_growth_curve()</a></code> function.</p></dd>


<dt id="arg-samp-scheme">samp_scheme<a class="anchor" aria-label="anchor" href="#arg-samp-scheme"></a></dt>
<dd><p>Tibble with columns <em>day</em> and n_sample_active. Rows only
need to contain the days on which sampling occurrs. Can be generated using
the <code><a href="define_sampling_scheme.html">define_sampling_scheme()</a></code> function.</p></dd>


<dt id="arg-founder-seqs">founder_seqs<a class="anchor" aria-label="anchor" href="#arg-founder-seqs"></a></dt>
<dd><p>Founder sequence(s) as a character string or a vector of
character strings, for example 'ACATG'. The founder sequence(s) may only
contain the characters ACGT, and no gaps are allowed. When modeling immune
fitness, they are expected to be codon-aligned.</p></dd>


<dt id="arg-generation-time">generation_time<a class="anchor" aria-label="anchor" href="#arg-generation-time"></a></dt>
<dd><p>Amount of time in days it takes a virus to complete
one full life cycle, from infecting one cell to exiting the cell and
infecting the next one (default: 1 day). Any inputs that are in days
will be converted to generations using this number.</p></dd>


<dt id="arg-mut-rate">mut_rate<a class="anchor" aria-label="anchor" href="#arg-mut-rate"></a></dt>
<dd><p>Mutation rate per-site, per-<em>generation</em> (default: 3.0e-5)</p></dd>


<dt id="arg-q">q<a class="anchor" aria-label="anchor" href="#arg-q"></a></dt>
<dd><p>Nucleotide substitution rate matrix Q with rows and columns named as
the nucleotides ACGT. Rows are from, columns are to. Can be generated using
the <code><a href="estimate_q.html">estimate_q()</a></code> function. The default is to calculate the Q matrix using
estimates of per-day rates from nearly neutral sites:
<code>wavess::calc_q_from_rates(wavess::hiv_mut_rates,mut_rate,generation_time)</code>.</p></dd>


<dt id="arg-recomb-rate">recomb_rate<a class="anchor" aria-label="anchor" href="#arg-recomb-rate"></a></dt>
<dd><p>Recombination rate per-site, per-<em>generation</em> (default:
1.4e-5)</p></dd>


<dt id="arg-act-to-lat">act_to_lat<a class="anchor" aria-label="anchor" href="#arg-act-to-lat"></a></dt>
<dd><p>Per-<em>day</em> rate that an active cell becomes latent (default:
0.001). Set this to 0 if you don't want to model latent cell dynamics.</p></dd>


<dt id="arg-lat-to-act">lat_to_act<a class="anchor" aria-label="anchor" href="#arg-lat-to-act"></a></dt>
<dd><p>Per-<em>day</em> rate that a latent cell becomes active (default:
0.01)</p></dd>


<dt id="arg-lat-prolif">lat_prolif<a class="anchor" aria-label="anchor" href="#arg-lat-prolif"></a></dt>
<dd><p>Per-<em>day</em> rate that a latent cell proliferates (default:
0.01)</p></dd>


<dt id="arg-lat-die">lat_die<a class="anchor" aria-label="anchor" href="#arg-lat-die"></a></dt>
<dd><p>Per-<em>day</em> rate that a latent cell dies (default: 0.01)</p></dd>


<dt id="arg-conserved-sites">conserved_sites<a class="anchor" aria-label="anchor" href="#arg-conserved-sites"></a></dt>
<dd><p>Vector of conserved bases named by position in the
founder sequence (indexed at 0). This can be generated using the
<code><a href="identify_conserved_sites.html">identify_conserved_sites()</a></code> function (default: NULL, i.e. no conserved
sites fitness costs)</p></dd>


<dt id="arg-conserved-cost">conserved_cost<a class="anchor" aria-label="anchor" href="#arg-conserved-cost"></a></dt>
<dd><p>Cost of mutation at conserved site, must be in the
range [0,1) where 0 indicates no cost. 1, which indicates no ability to
survive, is not allowed (default: 0.99)</p></dd>


<dt id="arg-ref-seq">ref_seq<a class="anchor" aria-label="anchor" href="#arg-ref-seq"></a></dt>
<dd><p>Reference sequence as a character string, which denotes the
"most fit" virus from a replicative perspective. A consensus sequence, that
can be used as the reference sequence, can be generated using the function
<code><a href="identify_conserved_sites.html">identify_conserved_sites()</a></code> (default: NULL, i.e. no fitness cost relative
to a reference sequence)</p></dd>


<dt id="arg-replicative-cost">replicative_cost<a class="anchor" aria-label="anchor" href="#arg-replicative-cost"></a></dt>
<dd><p>Replicative fitness cost, only relevant when ref_seq
is not NULL, must be in the range [0,1) where 0 indicates no cost. 1, which
indicates no ability to survive, is not allowed (default: 0.001)</p></dd>


<dt id="arg-epitope-locations">epitope_locations<a class="anchor" aria-label="anchor" href="#arg-epitope-locations"></a></dt>
<dd><p>Tibble of epitope locations and maximum fitness
costs with columns epi_start_nt, epi_end_nt, max_fitness_cost. These
epitopes are expected to be indexed at 0 and in a protein in the correct
reading frame, as the nucleotide sequences are translated to amino acids to
calculate the immune fitness cost. The maximum fitness cost must be in the
range [0,1) where 0 indicates no cost. 1, which indicates no ability to
survive. This epitope location tibble can be generated using the functions
<code><a href="get_epitope_frequencies.html">get_epitope_frequencies()</a></code> and <code><a href="sample_epitopes.html">sample_epitopes()</a></code>. (default: NULL, i.e.
no immune fitness costs)</p></dd>


<dt id="arg-n-for-imm">n_for_imm<a class="anchor" aria-label="anchor" href="#arg-n-for-imm"></a></dt>
<dd><p>Number of infected cells that must contain a given sequence
for that sequence to be recognized by the immune system, only relevant when
epitope_locations is not NULL (default: 100).</p></dd>


<dt id="arg-days-full-potency">days_full_potency<a class="anchor" aria-label="anchor" href="#arg-days-full-potency"></a></dt>
<dd><p>Number of <em>days</em> it takes for an immune response to
an epitope to reach full potency, only relevant when epitope_locations is
not NULL (default: 90).</p></dd>


<dt id="arg-immune-start-day">immune_start_day<a class="anchor" aria-label="anchor" href="#arg-immune-start-day"></a></dt>
<dd><p><em>Day</em> to start checking for an immune response, only
relevant when epitope_locations is not NULL (default: 0, but note that the
immune response will not actually start until there are at least
<code>n_for_imm</code> cells in the active population).</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Optional seed (default: NULL)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>List including: tibble of counts and mean fitness values, an alignment
of sampled sequences, and fitness of the sampled sequences. If latent cells
are sampled, then an alignment of the sampled latent cells will also be
returned.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Also note that some of the inputs are expected to be in units of
<em>generations</em> and some are expected to be in units of <em>days</em>. These choices
were made based on what empirical estimates are most often estimated present
in the literature. We have highlighted in the parameter descriptions which
inputs are which.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="fu">run_wavess</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="define_growth_curve.html">define_growth_curve</a></span><span class="op">(</span>n_gen <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  <span class="fu"><a href="define_sampling_scheme.html">define_sampling_scheme</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    sampling_frequency_active <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>    sampling_frequency_latent <span class="op">=</span> <span class="fl">10</span>, n_days <span class="op">=</span> <span class="fl">50</span></span></span>
<span class="r-in"><span>  <span class="op">)</span>,</span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"ATCG"</span>, <span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Zena Lapp, Narmada Sambaturu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

