<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Prepare input data • wavess</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Prepare input data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">wavess</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/prepare_input_data.html">Prepare input data</a></li>
    <li><a class="dropdown-item" href="../articles/run_wavess.html">Simulate within-host evolution with `wavess`</a></li>
    <li><a class="dropdown-item" href="../articles/analyze_output.html">Analyze wavess output</a></li>
    <li><a class="dropdown-item" href="../articles/python.html">Running wavess in Python</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Prepare input data</h1>
            
      

      <div class="d-none name"><code>prepare_input_data.Rmd</code></div>
    </div>

    
    
<p>wavess is a agent-based discrete-time within-host evolution
simulator. We assume that all virus life cycles within the infection are
synchronized into generations. Each generation is one full virus life
cycle, from infecting one cell to exiting the cell and infecting the
next one.</p>
<p>This vignette describes how to generate input data for running wavess
using functions from the package, as summarized in the table below.
Please note that you may also generate the input data on your own, which
may be necessary if your inputs require more customization than these
functions provide. However, we expect these functions to be sufficient
for most users.</p>
<table class="table">
<colgroup>
<col width="21%">
<col width="26%">
<col width="52%">
</colgroup>
<thead><tr class="header">
<th>
<code><a href="../reference/run_wavess.html">run_wavess()</a></code> argument</th>
<th>
<code>wavess</code> function to generate input</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>inf_pop_size</code></td>
<td><code><a href="../reference/define_growth_curve.html">define_growth_curve()</a></code></td>
<td><a href="#define-active-cell-growth">Define active cell
growth</a></td>
</tr>
<tr class="even">
<td><code>samp_scheme</code></td>
<td><code>define_samp_scheme()</code></td>
<td><a href="#define-sampling-scheme">Define sampling scheme</a></td>
</tr>
<tr class="odd">
<td><code>founder_seqs</code></td>
<td><code><a href="../reference/extract_seqs.html">extract_seqs()</a></code></td>
<td><a href="#extract-founder-sequence-from-an-alignment">Extract
founder sequence from an alignment</a></td>
</tr>
<tr class="even">
<td><code>q</code></td>
<td><code><a href="../reference/estimate_q.html">estimate_q()</a></code></td>
<td><a href="#estimate-q-matrix">Determine nucleotide substitution
probabilities</a></td>
</tr>
<tr class="odd">
<td><code>prob_*</code></td>
<td><code>rate_to_prob()</code></td>
<td><a href="#convert-rate-to-probability">Convert rate to
probability</a></td>
</tr>
<tr class="even">
<td><code>conserved_sites</code></td>
<td><code><a href="../reference/identify_conserved_sites.html">identify_conserved_sites()</a></code></td>
<td><a href="#identify-conserved-sites">Identify conserved
sites</a></td>
</tr>
<tr class="odd">
<td><code>ref_seq</code></td>
<td>
<code><a href="../reference/identify_conserved_sites.html">identify_conserved_sites()</a></code>,
<code><a href="../reference/extract_seqs.html">extract_seqs()</a></code>
</td>
<td><a href="#find-a-consensus-sequence">Get reference sequence that is
considered to be the best replicator</a></td>
</tr>
<tr class="even">
<td><code>epitope_locations</code></td>
<td>
<code><a href="../reference/get_epitope_frequencies.html">get_epitope_frequencies()</a></code>,
<code><a href="../reference/sample_epitopes.html">sample_epitopes()</a></code>
</td>
<td><a href="#sample-epitopes">Sample epitopes</a></td>
</tr>
</tbody>
</table>
<p>Each of the sections below goes into more detail.</p>
<p>We also provide examples of how to save the data. This can be useful
if generating the input data takes a while to run or if, rather than
using the built-in <code><a href="../reference/run_wavess.html">run_wavess()</a></code> function, you’d rather run
it using a python command line script (see
<code><a href="../articles/python.html">vignette("python")</a></code> for details). We save the files to a
directory called <code>input_data</code>. If you would like to do the
same, you will have to create that directory before running the file
saving commands. For example:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"input_data"</span><span class="op">)</span></span></code></pre></div>
<p>In each relevant section, we also provide examples of how to
visualize the data.</p>
<div class="section level2">
<h2 id="install-and-load-wavess">Install and load <code>wavess</code><a class="anchor" aria-label="anchor" href="#install-and-load-wavess"></a>
</h2>
<p>First, you need to install (if needed) and load the
<code>wavess</code> library, as well as a few others, and we will set
the plotting theme and a seed:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("remotes")</span></span>
<span><span class="co"># remotes::install_github("MolEvolEpid/wavess")</span></span>
<span><span class="co"># install.packages("ggplot2")</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://molevolepid.github.io/wavess/">wavess</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/emmanuelparadis/ape" class="external-link">ape</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'ape'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     where</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set seed</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span></code></pre></div>
<p>This will also load some example data that we will be using in this
vignette. All of this example data is related to the HIV gp120 protein.
Please note that the default parameters for <code><a href="../reference/run_wavess.html">run_wavess()</a></code>
are also based on this example.</p>
<p>In particular, we include the following example data in the
package:</p>
<ul>
<li>
<code>hxb2_cons_founder</code>: an alignment in
<code><a href="https://rdrr.io/pkg/ape/man/DNAbin.html" class="external-link">ape::DNAbin</a></code> format that includes the HIV full-genome
sequence for the HXB2 reference sequence, the consensus sequence, and an
example founder sequence.</li>
<li>
<code>hiv_mut_rates</code>: per-site per-day rates of change between
specific nucleotides</li>
<li>
<code>hiv_env_flt_2022</code>: the first 10 sequences from the
filtered HIV ENV alignment downloaded from the <a href="https://www.hiv.lanl.gov/content/sequence/NEWALIGN/align.html" class="external-link">LANL
HIV sequence database</a>, in <code><a href="https://rdrr.io/pkg/ape/man/DNAbin.html" class="external-link">ape::DNAbin</a></code> format. Feel free
to download the entire alignment if you’d like.</li>
<li>
<code>conserved_sites</code>: a vector of conserved sites for the
example founder sequence. More details on this below.</li>
<li>
<code>env_features</code>: a tibble of binding, contact, and and
neutralization features for HIV ENV from the <a href="https://www.hiv.lanl.gov/components/sequence/HIV/neutralization/download_db.comp" class="external-link">LANL
HIV immunology database</a>.</li>
</ul>
<p>You can check out the documentation for these datasets for more
details.</p>
<p>If you want to load your own data, you can use
<code><a href="https://rdrr.io/pkg/ape/man/read.dna.html" class="external-link">ape::read.dna()</a></code> or <code><a href="https://rdrr.io/pkg/ape/man/read.dna.html" class="external-link">ape::read.FASTA()</a></code> to load
alignment files and <code><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">readr::read_delim()</a></code> to load delimited
files.</p>
</div>
<div class="section level2">
<h2 id="required-run_wavess-inputs">Required <code>run_wavess()</code> inputs<a class="anchor" aria-label="anchor" href="#required-run_wavess-inputs"></a>
</h2>
<p>The three required inputs to <code>run_waves()</code> are a data
frame of the infected cell population size for each <em>generation</em>,
a data frame including the <em>days</em> on which to sample and how many
sequences to sample (<code>samp_scheme</code>), and a vector of the
founder sequence(s) in character string format
(<code>founder_seqs</code>). In this section, we will introduce
functions that can help generate these inputs.</p>
<p>Note that, in general, some of the inputs are in units of generations
and some are in units of days. We made these decisions mainly based on
what information is most readily available in the literature related to
each parameter value.</p>
<div class="section level3">
<h3 id="define-active-cell">Define active cell growth<a class="anchor" aria-label="anchor" href="#define-active-cell"></a>
</h3>
<p><code><a href="../reference/run_wavess.html">run_wavess()</a></code> requires a <code>inf_pop_size</code> data
frame as input where each row is a generation with the following
columns:</p>
<ul>
<li>
<code>generation</code>: Each generation to be simulated (must be
consecutive whole numbers starting with 0)</li>
<li>
<code>active_cell_count</code>: Number of active infected cells in
each generation</li>
</ul>
<p>You can create this data frame yourself, but we also provide a helper
function to generate it: <code><a href="../reference/define_growth_curve.html">define_growth_curve()</a></code>.</p>
<p>All of the arguments in this function have defaults, so you can
run:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/define_growth_curve.html">define_growth_curve</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 5,001 × 2</span></span></span>
<span><span class="co">#&gt;    generation active_cell_count</span></span>
<span><span class="co">#&gt;         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>          0                10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>          1                13</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>          2                17</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>          3                22</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>          4                29</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>          5                37</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>          6                48</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>          7                62</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>          8                80</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>          9               103</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 4,991 more rows</span></span></span></code></pre></div>
<p>This will generate an infected cell population size following a
logistic growth curve. By default, the starting population size
(<code>n0</code>) is 10, the carrying capacity (<code>carry_cap</code>)
is 2000, the maximum growth rate (<code>max_growth_rate</code>) is 0.3,
and the simulation is run for a maximum 5000 generations
(<code>n_gen</code>). All of these defaults can be changed by altering
the input parameters of the function.</p>
<p>Here is an example where we allow the simulation to run for longer
(for as long as sequences are being sampled):</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">inf_pop_size</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/define_growth_curve.html">define_growth_curve</a></span><span class="op">(</span>n_gen <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 10,001 × 2</span></span></span>
<span><span class="co">#&gt;    generation active_cell_count</span></span>
<span><span class="co">#&gt;         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>          0                10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>          1                13</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>          2                17</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>          3                22</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>          4                29</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>          5                37</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>          6                48</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>          7                62</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>          8                80</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>          9               103</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 9,991 more rows</span></span></span></code></pre></div>
<p>If you’d like to write this to a file:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_csv</a></span><span class="op">(</span><span class="va">inf_pop_size</span>, <span class="st">"input_data/inf_pop_size.csv"</span><span class="op">)</span></span></code></pre></div>
<p>A couple notes:</p>
<ul>
<li>The number of founders must equal the initial population size.</li>
<li>Note that if you start with multiple different founder sequences,
you may stochastically lose some of those sequences after generation
<ol start="0" style="list-style-type: decimal"><li>
</li></ol>
</li>
</ul>
<p>You can visualize the active cell dynamics over time using the
following code:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot active cell counts</span></span>
<span><span class="va">inf_pop_size</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">generation</span> <span class="op">&lt;=</span> <span class="fl">500</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">generation</span>, y <span class="op">=</span> <span class="va">active_cell_count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Generation"</span>, y <span class="op">=</span> <span class="st">"Number of infected active cells"</span><span class="op">)</span></span></code></pre></div>
<p><img src="prepare_input_data_files/figure-html/plot_counts-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="define-sampling-scheme">Define sampling scheme<a class="anchor" aria-label="anchor" href="#define-sampling-scheme"></a>
</h3>
<p><code><a href="../reference/run_wavess.html">run_wavess()</a></code> requires a <code>samp_scheme</code> data
frame as input where each row is a <em>day</em> with the following
columns:</p>
<ul>
<li>
<code>day</code>: Each day on which to sample sequences</li>
<li>
<code>n_sample_active</code>: Number of sequences from active cells
to sample for that day (note that in the simulation output, this number
may be lower if the population size is smaller than the requested number
of sequences to sample)</li>
<li>
<code>n_sample_latent</code>: Number of sequences from latent cells
to sample for that day (note that in the simulation output, this number
may be lower if the population size is smaller than the requested number
of sequences to sample)</li>
</ul>
<p>You can create this data frame yourself, but we also provide a helper
function to generate it: <code><a href="../reference/define_sampling_scheme.html">define_sampling_scheme()</a></code>.</p>
<p>All of the arguments in this function have defaults, so you can
run:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/define_sampling_scheme.html">define_sampling_scheme</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 11 × 3</span></span></span>
<span><span class="co">#&gt;      day n_sample_active n_sample_latent</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     0              20              20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>   365              20              20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>   730              20              20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>  <span style="text-decoration: underline;">1</span>095              20              20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>  <span style="text-decoration: underline;">1</span>460              20              20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>  <span style="text-decoration: underline;">1</span>825              20              20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>  <span style="text-decoration: underline;">2</span>190              20              20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>  <span style="text-decoration: underline;">2</span>555              20              20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>  <span style="text-decoration: underline;">2</span>920              20              20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>  <span style="text-decoration: underline;">3</span>285              20              20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span>  <span style="text-decoration: underline;">3</span>650              20              20</span></span></code></pre></div>
<p>This will lead to sampling a maximum of 20 sequences
(<code>max_samp</code>) every 365 days (<code>sampling_frequency</code>)
for 3650 days (<code>n_days</code>).</p>
<p>Here is an example with sampling over a shorter time period, but more
frequent sampling and fewer samples taken at each sampling event:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">samp_scheme</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/define_sampling_scheme.html">define_sampling_scheme</a></span><span class="op">(</span></span>
<span>  sampling_frequency_active <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  max_samp_active <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  n_days <span class="op">=</span> <span class="fl">365</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 14 × 3</span></span></span>
<span><span class="co">#&gt;      day n_sample_active n_sample_latent</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     0              10              20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>    30              10               0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>    60              10               0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>    90              10               0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>   120              10               0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>   150              10               0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>   180              10               0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>   210              10               0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>   240              10               0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>   270              10               0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span>   300              10               0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span>   330              10               0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">13</span>   360              10               0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">14</span>   365               0              20</span></span></code></pre></div>
<p>Note that the simulation will automatically end after the last
sampling time, even if the growth curve continues for more
generations.</p>
<p>If you’d like to write this to a file:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_csv</a></span><span class="op">(</span><span class="va">samp_scheme</span>, <span class="st">"input_data/samp_scheme.csv"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="extract-founder-sequence-from-an-alignment">Extract founder sequence from an alignment<a class="anchor" aria-label="anchor" href="#extract-founder-sequence-from-an-alignment"></a>
</h3>
<p><code><a href="../reference/run_wavess.html">run_wavess()</a></code> takes as input a character vector of
founder sequence(s) (<code>founder_seqs</code>). All sequences must be
the same length and not contain gaps.</p>
<p>We provide the function <code><a href="../reference/extract_seqs.html">extract_seqs()</a></code> to extract a
founder sequence from an alignment in <code><a href="https://rdrr.io/pkg/ape/man/DNAbin.html" class="external-link">ape::DNAbin</a></code> format
and convert it into a character vector. The alignment can be read in
using <code><a href="https://rdrr.io/pkg/ape/man/read.dna.html" class="external-link">ape::read.FASTA()</a></code> or
<code><a href="https://rdrr.io/pkg/ape/man/read.dna.html" class="external-link">ape::read.dna()</a></code>.</p>
<p>NOTE: We do not recommend using this function to extract more than
one founder sequence to initiate a single simulation because if any gaps
are present then the founder sequences will not be the same length,
which will lead to an error. Rather, if you would like to simulate
multiple founders, we recommend aligning all the founder sequences and
then stripping gaps such that the alignment remains codon-aligned,
especially if you plan to simulate immune fitness because epitopes are
translated to amino acids to calculate immune fitness costs.</p>
<p>This alignment is of the entire HIV genome, but we’re only interested
in the ENV gp120 gene, which we can subset to using the
<code>start</code> and <code>end</code> arguments.</p>
<p>To do this, we first have to know what the start and end positions of
gp120 are in the alignment. We know the start and end positions of gp120
in HXB2 from <a href="https://www.hiv.lanl.gov/content/sequence/HIV/MAP/landmark.html" class="external-link">here</a>
(6225 and 7757). To get the start and end positions in our alignment, we
can therefore find what alignment positions correspond to these HXB2
coordinates:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gp120_start</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span></span>
<span>  <span class="va">hxb2_cons_founder</span><span class="op">[</span><span class="st">"B.FR.83.HXB2_LAI_IIIB_BRU.K03455"</span>, <span class="op">]</span></span>
<span><span class="op">)</span> <span class="op">!=</span> <span class="st">"-"</span><span class="op">)</span> <span class="op">==</span> <span class="fl">6225</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">gp120_end</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span></span>
<span>  <span class="va">hxb2_cons_founder</span><span class="op">[</span><span class="st">"B.FR.83.HXB2_LAI_IIIB_BRU.K03455"</span>, <span class="op">]</span></span>
<span><span class="op">)</span> <span class="op">!=</span> <span class="st">"-"</span><span class="op">)</span> <span class="op">==</span> <span class="fl">7757</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre></div>
<p>Then we can use these start and end positions to extract gp120 for
our founder sequence:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">founder_seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_seqs.html">extract_seqs</a></span><span class="op">(</span><span class="va">hxb2_cons_founder</span>, <span class="st">"B.US.2011.DEMB11US006.KC473833"</span>,</span>
<span>  start <span class="op">=</span> <span class="va">gp120_start</span>, end <span class="op">=</span> <span class="va">gp120_end</span></span>
<span><span class="op">)</span><span class="op">$</span><span class="va">founder</span></span></code></pre></div>
<p>This function can also take a reference sequence name
(<code>ref_name</code>), which can be used as the input to the
<code>ref_seq</code> argument in <code><a href="../reference/run_wavess.html">run_wavess()</a></code>. More on this
later.</p>
<p>The easiest way to save this sequence as a fasta file is to convert
it into an <code><a href="https://rdrr.io/pkg/ape/man/DNAbin.html" class="external-link">ape::DNAbin</a></code> object and then save it:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">founder_seq</span>, <span class="st">""</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ape/man/as.alignment.html" class="external-link">as.DNAbin</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ape/man/write.dna.html" class="external-link">write.FASTA</a></span><span class="op">(</span><span class="st">"input_data/founder.fasta"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="estimate-q-matrix">Estimate Q matrix<a class="anchor" aria-label="anchor" href="#estimate-q-matrix"></a>
</h3>
<p>The model requires as input a Q matrix (<code>q</code>), which
defines the rates of nucleotide substitution for each possible
transition between nucleotides. By default, this matrix is estimated
from approximately neutral sites:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">hiv_q_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_q_from_rates.html">calc_q_from_rates</a></span><span class="op">(</span><span class="va">hiv_mut_rates</span>, mut_rate <span class="op">=</span> <span class="fl">2.4e-5</span>, generation_time <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;            A          C          G          T</span></span>
<span><span class="co">#&gt; A -1.3935602  0.1650215  1.1002145  0.1283242</span></span>
<span><span class="co">#&gt; C  0.9167813 -3.2089352  0.0915972  2.2005567</span></span>
<span><span class="co">#&gt; G  2.9338881  0.0182889 -3.3189236  0.3667466</span></span>
<span><span class="co">#&gt; T  0.5499598  1.8338005  0.5501132 -2.9338736</span></span></code></pre></div>
<p>This matrix is based on the rates from <a href="https://doi.org/10.1093/ve/vex003" class="external-link">Zanini et al. 2017</a>, and is
converted to an approximate Q matrix by dividing by the estimated
overall nucleotide substitution rate.</p>
<p><strong>If you are simulating an organism other than HIV, you will
want to provide a Q matrix estimated for that organism.</strong> Any
phylogenetic tree building program will estimate a Q matrix. For
instance, the Q matrix is returned when estimating a tree using <a href="http://www.iqtree.org/" class="external-link">IQ-TREE</a> (in the .iqtree output file).
We also provide the function <code><a href="../reference/estimate_q.html">estimate_q()</a></code> to estimate Q
from a multiple sequence alignment (in <code><a href="https://rdrr.io/pkg/ape/man/DNAbin.html" class="external-link">ape::DNAbin</a></code> format)
directly in R. Ideally, the input for this function would be an
alignment of sequences from within-host evolution of a representative
person. However, we have found that the simulation output is relatively
robust to variations in the Q matrix, so if you don’t have a within-host
alignment, then you can input an alignment with somewhat closely related
sequences (e.g. from the same subtype). The function outputs a Q matrix
where the rows are the “from” nucleotide and the columns are the “to”
nucleotide.</p>
<p>Here is an example (but note that this matrix isn’t accurate because
it’s simply a random set of sequences):</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/estimate_q.html">estimate_q</a></span><span class="op">(</span><span class="va">hiv_env_flt_2022</span><span class="op">)</span></span>
<span><span class="co">#&gt; optimize edge weights:  -13347 --&gt; -13272.97 </span></span>
<span><span class="co">#&gt; optimize rate matrix:  -13272.97 --&gt; -12963.97 </span></span>
<span><span class="co">#&gt; optimize invariant sites:  -12963.97 --&gt; -12460.48 </span></span>
<span><span class="co">#&gt; optimize free rate parameters:  -12460.48 --&gt; -10603 </span></span>
<span><span class="co">#&gt; optimize edge weights:  -10603 --&gt; -10590.63 </span></span>
<span><span class="co">#&gt; optimize rate matrix:  -10590.63 --&gt; -10581.46 </span></span>
<span><span class="co">#&gt; optimize invariant sites:  -10581.46 --&gt; -10581.46 </span></span>
<span><span class="co">#&gt; optimize free rate parameters:  -10581.46 --&gt; -10577.81 </span></span>
<span><span class="co">#&gt; optimize edge weights:  -10577.81 --&gt; -10577.48 </span></span>
<span><span class="co">#&gt; optimize rate matrix:  -10577.48 --&gt; -10577.26 </span></span>
<span><span class="co">#&gt; optimize invariant sites:  -10577.26 --&gt; -10577.26 </span></span>
<span><span class="co">#&gt; optimize free rate parameters:  -10577.26 --&gt; -10575.89 </span></span>
<span><span class="co">#&gt; optimize edge weights:  -10575.89 --&gt; -10575.83 </span></span>
<span><span class="co">#&gt; optimize rate matrix:  -10575.83 --&gt; -10575.79 </span></span>
<span><span class="co">#&gt; optimize invariant sites:  -10575.79 --&gt; -10575.79 </span></span>
<span><span class="co">#&gt; optimize free rate parameters:  -10575.79 --&gt; -10575.18 </span></span>
<span><span class="co">#&gt; optimize edge weights:  -10575.18 --&gt; -10575.16 </span></span>
<span><span class="co">#&gt; optimize rate matrix:  -10575.16 --&gt; -10575.14 </span></span>
<span><span class="co">#&gt; optimize invariant sites:  -10575.14 --&gt; -10575.14 </span></span>
<span><span class="co">#&gt; optimize free rate parameters:  -10575.14 --&gt; -10574.92 </span></span>
<span><span class="co">#&gt; optimize edge weights:  -10574.92 --&gt; -10574.91 </span></span>
<span><span class="co">#&gt; optimize rate matrix:  -10574.91 --&gt; -10574.91 </span></span>
<span><span class="co">#&gt; optimize invariant sites:  -10574.91 --&gt; -10574.91 </span></span>
<span><span class="co">#&gt; optimize free rate parameters:  -10574.91 --&gt; -10574.84 </span></span>
<span><span class="co">#&gt; optimize edge weights:  -10574.84 --&gt; -10574.84 </span></span>
<span><span class="co">#&gt; optimize rate matrix:  -10574.84 --&gt; -10574.83 </span></span>
<span><span class="co">#&gt; optimize invariant sites:  -10574.83 --&gt; -10574.83 </span></span>
<span><span class="co">#&gt; optimize free rate parameters:  -10574.83 --&gt; -10574.82 </span></span>
<span><span class="co">#&gt; optimize edge weights:  -10574.82 --&gt; -10574.81 </span></span>
<span><span class="co">#&gt; optimize rate matrix:  -10574.81 --&gt; -10574.81 </span></span>
<span><span class="co">#&gt; optimize invariant sites:  -10574.81 --&gt; -10574.81 </span></span>
<span><span class="co">#&gt; optimize free rate parameters:  -10574.81 --&gt; -10574.81 </span></span>
<span><span class="co">#&gt; optimize edge weights:  -10574.81 --&gt; -10574.81 </span></span>
<span><span class="co">#&gt; optimize rate matrix:  -10574.81 --&gt; -10574.81 </span></span>
<span><span class="co">#&gt; optimize invariant sites:  -10574.81 --&gt; -10574.81 </span></span>
<span><span class="co">#&gt; optimize free rate parameters:  -10574.81 --&gt; -10574.8 </span></span>
<span><span class="co">#&gt; optimize edge weights:  -10574.8 --&gt; -10574.8 </span></span>
<span><span class="co">#&gt; optimize rate matrix:  -10574.8 --&gt; -10574.8 </span></span>
<span><span class="co">#&gt; optimize invariant sites:  -10574.8 --&gt; -10574.8 </span></span>
<span><span class="co">#&gt; optimize free rate parameters:  -10574.8 --&gt; -10574.8 </span></span>
<span><span class="co">#&gt; optimize edge weights:  -10574.8 --&gt; -10574.8 </span></span>
<span><span class="co">#&gt; optimize rate matrix:  -10574.8 --&gt; -10574.8 </span></span>
<span><span class="co">#&gt; optimize invariant sites:  -10574.8 --&gt; -10574.8 </span></span>
<span><span class="co">#&gt; optimize free rate parameters:  -10574.8 --&gt; -10574.8 </span></span>
<span><span class="co">#&gt; optimize edge weights:  -10574.8 --&gt; -10574.8</span></span>
<span><span class="co">#&gt;            A          C          G          T</span></span>
<span><span class="co">#&gt; A -1.5940889  0.3894774  1.0395679  0.1650437</span></span>
<span><span class="co">#&gt; C  0.7692561 -2.0446348  0.1948555  1.0805232</span></span>
<span><span class="co">#&gt; G  1.5490726  0.1470087 -1.9302494  0.2341682</span></span>
<span><span class="co">#&gt; T  0.2459336  0.8152003  0.2341682 -1.2953020</span></span></code></pre></div>
<p>By default, a neighbor joining tree is created using
<code>ape::bionj(ape::dist.dna(aln, model = 'TN93'))</code>. Instead,
you can provide an input tree (<code>tr</code>). From this, the Q matrix
is computed with <code><a href="https://klausvigo.github.io/phangorn/reference/pml_bb.html" class="external-link">phangorn::pml_bb()</a></code> using by default a
GTR+I+R4 model of nucleotide substitution (<code>model</code>) and no
tree rearrangement (<code>rearrangement</code>).</p>
<p>To save this to a file:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">hiv_q_mat</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"nt_from"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_csv</a></span><span class="op">(</span><span class="st">"input_data/hiv_q_mat.csv"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Note:</strong> to read this back into R and get it into
matrix form, you can use this code:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"input_data/hiv_q_mat.csv"</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;            A          C          G          T</span></span>
<span><span class="co">#&gt; A -1.3935602  0.1650215  1.1002145  0.1283242</span></span>
<span><span class="co">#&gt; C  0.9167813 -3.2089352  0.0915972  2.2005567</span></span>
<span><span class="co">#&gt; G  2.9338881  0.0182889 -3.3189236  0.3667466</span></span>
<span><span class="co">#&gt; T  0.5499598  1.8338005  0.5501132 -2.9338736</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="visualize-latent-cell-dynamics">Latent cell dynamics<a class="anchor" aria-label="anchor" href="#visualize-latent-cell-dynamics"></a>
</h2>
<p>The required inputs related to latency are per-day rates that these
events happen to a cell. (Note that to “turn off” any of these, all you
have to do is set the rate equal to 0.) It is helpful to visualize the
latent cell dynamics prior to simulating data with a given set of rates.
While the way the model implements latency is stochastic, meaning that a
different number of cells will become latent in each simulation, the
overall trends should be similar across simulations, given the same set
of rates.</p>
<p>To make these plots, you need the active cell counts, which can be
generated from the <code><a href="../reference/define_growth_curve.html">define_growth_curve()</a></code> function described
above. We will therefore be using the <code>inf_pop_size</code> dataset
we created earlier in the vignette. Also, the rates related to latency
are converted to probabilities for simulation, so we will convert them
here before plotting.</p>
<p><strong>Please note that we assume the rates (and thus per-generation
probabilities) are small for all latency parameters</strong>, such that
it is unlikely that multiple events (activate, die, proliferate) will
occur to a single latent cell in a single [active cell] generation. If
multiple events occur to a single latent cell, then the first event in
this ordered list will be chosen as the event that occurred to the cell
in that generation: cell becomes active, cell dies, cell
proliferates.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set parameters to get latent curve</span></span>
<span><span class="va">to_latent</span> <span class="op">&lt;-</span> <span class="fl">0.001</span></span>
<span><span class="va">to_active</span> <span class="op">&lt;-</span> <span class="fl">0.01</span></span>
<span><span class="va">proliferation</span> <span class="op">&lt;-</span> <span class="fl">0.01</span></span>
<span><span class="va">death</span> <span class="op">&lt;-</span> <span class="fl">0.01</span></span>
<span></span>
<span><span class="co"># get latent cell count for each generation</span></span>
<span><span class="va">latent</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">active_latent_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">inf_pop_size</span><span class="op">$</span><span class="va">active_cell_count</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">n_to_latent</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">rbinom</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">x</span>, prob <span class="op">=</span> <span class="va">to_latent</span><span class="op">)</span></span>
<span>  <span class="va">to_active</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">rbinom</a></span><span class="op">(</span><span class="va">latent</span>, <span class="fl">1</span>, prob <span class="op">=</span> <span class="va">to_active</span><span class="op">)</span></span>
<span>  <span class="va">to_proliferate</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">rbinom</a></span><span class="op">(</span><span class="va">latent</span>, <span class="fl">1</span>, prob <span class="op">=</span> <span class="va">proliferation</span><span class="op">)</span></span>
<span>  <span class="va">to_die</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">rbinom</a></span><span class="op">(</span><span class="va">latent</span>, <span class="fl">1</span>, prob <span class="op">=</span> <span class="va">death</span><span class="op">)</span></span>
<span>  <span class="va">n_to_active</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">to_active</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="va">n_to_die</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">(</span><span class="va">to_die</span> <span class="op">-</span> <span class="va">to_active</span><span class="op">)</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="va">n_to_proliferate</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">(</span><span class="va">to_proliferate</span> <span class="op">-</span> <span class="va">to_die</span> <span class="op">-</span> <span class="va">to_active</span><span class="op">)</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>    Active <span class="op">=</span> <span class="va">x</span>, Latent <span class="op">=</span> <span class="va">latent</span>, <span class="va">n_to_latent</span>, <span class="va">n_to_active</span>,</span>
<span>    <span class="va">n_to_proliferate</span>, <span class="va">n_to_die</span></span>
<span>  <span class="op">)</span></span>
<span>  <span class="va">latent</span> <span class="op">&lt;&lt;-</span> <span class="va">latent</span> <span class="op">+</span> <span class="va">n_to_latent</span> <span class="op">-</span> <span class="va">n_to_active</span> <span class="op">+</span> <span class="va">n_to_proliferate</span> <span class="op">-</span> <span class="va">n_to_die</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>gen <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/row_number.html" class="external-link">row_number</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot (latent) cell counts</span></span>
<span><span class="va">active_latent_counts</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gen</span> <span class="op">&lt;=</span> <span class="fl">300</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">Active</span>, <span class="va">Latent</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">gen</span>, y <span class="op">=</span> <span class="va">value</span>, col <span class="op">=</span> <span class="va">name</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># scale_y_log10() + # uncomment this if you want to scale the y axis by log10</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Generation"</span>, y <span class="op">=</span> <span class="st">"Number of infected cells"</span>, col <span class="op">=</span> <span class="st">"Type of cell"</span><span class="op">)</span></span></code></pre></div>
<p><img src="prepare_input_data_files/figure-html/viz_latent-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="selection">Selection<a class="anchor" aria-label="anchor" href="#selection"></a>
</h2>
<p><code>wavess</code> can simulate three types of selection: conserved
sites, replicative fitness relative to a reference sequence, and B-cell
immune selection at user-defined epitopes. By default, no selection is
included in the simulations. However, we recommend including selection
to obtain realistic model outputs.</p>
<p>Below, we describe how to generate the inputs for each of these
selective pressures.</p>
<div class="section level3">
<h3 id="fitness-costs">Fitness costs<a class="anchor" aria-label="anchor" href="#fitness-costs"></a>
</h3>
<p>Each of the different forms of selection has a fitness cost
associated with it. The fitness cost can be in the range [0,1), where 0
indicates no cost. 1, which indicates no ability to survive, is not
allowed as we assume that we are only simulating scenarios where
infection is established and the viruses are at least somewhat viable.
However, using a very high fitness cost, e.g. of 0.99, effectively
simulates purifying selection due to the compounding nature of the
fitness cost over many generations.</p>
<p>The fitness, F, of a virus is defined by the product of the fitness
of each component:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>F</mi><mo>=</mo><msub><mi>F</mi><mi>C</mi></msub><mo>*</mo><msub><mi>F</mi><mi>R</mi></msub><mo>*</mo><msub><mi>F</mi><mi>I</mi></msub></mrow><annotation encoding="application/x-tex">F=F_C*F_R*F_I</annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>F</mi><mi>C</mi></msub><annotation encoding="application/x-tex">F_C</annotation></semantics></math>
is the conserved fitness,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>F</mi><mi>R</mi></msub><annotation encoding="application/x-tex">F_R</annotation></semantics></math>
is the replicative fitness, and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>F</mi><mi>I</mi></msub><annotation encoding="application/x-tex">F_I</annotation></semantics></math>
is the immune fitness.</p>
<p>The equation used to compute
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>F</mi><mi>C</mi></msub><annotation encoding="application/x-tex">F_C</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>F</mi><mi>R</mi></msub><annotation encoding="application/x-tex">F_R</annotation></semantics></math>
is:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>F</mi><mrow><mi>C</mi><mi>/</mi><mi>R</mi></mrow></msub><mo>=</mo><msup><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><mi>c</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>n</mi></msup></mrow><annotation encoding="application/x-tex">F_{C/R} = (1-c)^n</annotation></semantics></math>
Where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>c</mi><annotation encoding="application/x-tex">c</annotation></semantics></math>
is either the conserved fitness cost or the replicative fitness cost,
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>n</mi><annotation encoding="application/x-tex">n</annotation></semantics></math>
is the number of mutations at conserved sites or the number of sites
that differ from the reference sequence, respectively.</p>
<p>When both forms of fitness are used, if the position is considered to
be a conserved site, then it is not considered for replicative
fitness.</p>
<p>Here is a plot of how the cost and number of mutations influences
overall fitness:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">costs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.001</span>, <span class="fl">0.01</span>, <span class="fl">0.1</span>, <span class="fl">0.99</span><span class="op">)</span></span>
<span><span class="va">n_muts</span> <span class="op">&lt;-</span> <span class="fl">0</span><span class="op">:</span><span class="fl">100</span></span>
<span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>n_mut <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">n_muts</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">costs</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, cost <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">costs</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">n_muts</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>fitness <span class="op">=</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">cost</span><span class="op">)</span><span class="op">**</span><span class="va">n_mut</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">n_mut</span>, y <span class="op">=</span> <span class="va">fitness</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">cost</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Number of mutations"</span>, y <span class="op">=</span> <span class="st">"Fitness"</span>, col <span class="op">=</span> <span class="st">"Cost per mutation"</span><span class="op">)</span></span></code></pre></div>
<p><img src="prepare_input_data_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>The calculation of immune fitness is a bit more involved, but the
result is that it is computed as the largest fitness cost across all
immune-recognized epitopes:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>F</mi><mi>I</mi></msub><mo>=</mo><mi>m</mi><mi>a</mi><mi>x</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>c</mi><mrow><mi>e</mi><mi>p</mi><mi>i</mi><mi>t</mi><mi>o</mi><mi>p</mi><mi>e</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">F_I = max(c_{epitope})</annotation></semantics></math>
Where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>c</mi><mrow><mi>e</mi><mi>p</mi><mi>i</mi><mi>t</mi><mi>o</mi><mi>p</mi><mi>e</mi></mrow></msub><annotation encoding="application/x-tex">c_{epitope}</annotation></semantics></math>
is a vector of the costs of each epitope recognized by the immune
system. Note that this cost changes over generations. Please refer to
the manuscript for more details.</p>
<p>In the next sections, we will discuss how to identify conserved
sites, create a reference sequence, and generate epitopes.</p>
</div>
<div class="section level3">
<h3 id="an-important-note-about-indexing-and-reference-sequences">An important note about indexing and reference sequences<a class="anchor" aria-label="anchor" href="#an-important-note-about-indexing-and-reference-sequences"></a>
</h3>
<p>First, all nucleotide positions must be indexed relative to the start
of the founder sequence in the simulation. Since the back-end is
implemented in Python, we expect the indexing to begin at 0. This is
because you can run wavess using the <code><a href="../reference/run_wavess.html">run_wavess()</a></code> function
in R, but also directly from Python (see <code><a href="../articles/python.html">vignette("python")</a></code>
for details). This way, the inputs are the same regardless of what
program you use to run wavess. This is relevant for conserved sites and
epitopes.</p>
<p>Second, information of interest such as conserved sites or antibody
contacts, is often computed or provided relative to what is called a
reference sequence. For HIV, the reference sequence is usually HXB2.
<strong>This community reference sequence is different than how we
define a reference sequence in <code>wavess</code>.</strong> The
community reference sequence (e.g. HXB2) is a standard that researchers
use to more easily be able to share and compare information. The
reference sequence defined in <code>wavess</code> is one that is
believed to be representative of the “most fit” sequence (when not under
immune pressure), however you would like to think of it.</p>
<p>When information of interest is provided relative to the community
reference sequence, we must change the indexing of the information to be
relative to the founder sequence that will be used in the
simulation.</p>
<p>The easiest way to ensure that the indexing is correct is to have an
alignment consisting of the founder sequence, the community reference
sequence (if needed), and the wavess reference sequence (if being used),
where the start and end of the alignment are the start and end of the
founder sequence to be used in the simulation.</p>
<p>If you have an alignment with sequences that are longer than the
input sequence you want to simulate, you can use the
<code><a href="../reference/slice_aln.html">slice_aln()</a></code> function to slice out the desired section of
the alignment. Here’s an example:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">gp120</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slice_aln.html">slice_aln</a></span><span class="op">(</span><span class="va">hxb2_cons_founder</span>, <span class="va">gp120_start</span>, <span class="va">gp120_end</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; 3 DNA sequences in binary format stored in a matrix.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; All sequences of same length: 1563 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Labels:</span></span>
<span><span class="co">#&gt; B.FR.83.HXB2_LAI_IIIB_BRU.K03455</span></span>
<span><span class="co">#&gt; CON_B(1295)</span></span>
<span><span class="co">#&gt; B.US.2011.DEMB11US006.KC473833</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Base composition:</span></span>
<span><span class="co">#&gt;     a     c     g     t </span></span>
<span><span class="co">#&gt; 0.381 0.160 0.214 0.245 </span></span>
<span><span class="co">#&gt; (Total: 4.69 kb)</span></span></code></pre></div>
<p>We will use this alignment below.</p>
</div>
<div class="section level3">
<h3 id="identify-conserved-sites">Identify conserved sites<a class="anchor" aria-label="anchor" href="#identify-conserved-sites"></a>
</h3>
<p>In <code><a href="../reference/run_wavess.html">run_wavess()</a></code> you can provide a vector of conserved
sites and the conserved nuceltoide (<code>conserved_sites</code>
argument). Mutations away from a conserved nucleotide have a fitness
cost defined by the <code>conserved_cost</code> argument.</p>
<p>To generate a vector of conserved sites from an alignment, you can
use the function <code><a href="../reference/identify_conserved_sites.html">identify_conserved_sites()</a></code>, which outputs
a data frame with 5 columns:</p>
<ul>
<li>
<code>founder_pos</code>: the position in the founder</li>
<li>
<code>founder_base</code>: the base in the founder sequence</li>
<li>
<code>consensus_base</code>: the consensus base</li>
<li>
<code>consensus_prop</code>: the proportion of sequences that have
the consensus base</li>
<li>
<code>conserved</code>: whether (‘Yes’) or not (‘No’) the site is
considered conserved, based on the threshold value defined by the
<code>thresh</code> argument (default: 0.99)</li>
</ul>
<p>You can use this function in two ways. If you have an alignment that
includes the founder sequence of interest, all you need to provide is
the alignment in <code><a href="https://rdrr.io/pkg/ape/man/DNAbin.html" class="external-link">ape::DNAbin</a></code> format and the name of the
founder sequence in the alignment. Note that the function assumes that
the alignment consists of only the segment of the founder sequence that
you want to simulate (i.e., the beginning of the alignment is the
beginning of the founder sequence that you want to simulate, and the end
of the alignment is the end of the sequence you want to simulate).</p>
<p>We will be using the built-in <code>hiv_env_flt_2022</code> alignment
as an example. However, this contains only 10 sequences. <strong>In
reality, you should use many more sequences.</strong> For HIV, you can
download the entire filtered alignment for each gene or for the entire
genome from the <a href="https://www.hiv.lanl.gov/content/sequence/NEWALIGN/align.html" class="external-link">LANL
HIV sequence database</a>.</p>
<p>For the example founder gp120 sequence we use in the package
(B.US.2011.DEMB11US006.KC473833), we also provide a vector of conserved
sites for this founder sequence (<code>founder_conserved_sites</code>)
for ease of use.</p>
<p>First, since the <code>hxb2_cons_founder</code> sequence is all of
<em>env</em>, but we just want to simulate gp120, we have to slice out
that part of the alignment:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get hxb2 gp120 length and end position in flt alignment</span></span>
<span><span class="va">len_hxb2_gp120</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="fu"><a href="../reference/extract_seqs.html">extract_seqs</a></span><span class="op">(</span><span class="va">hxb2_cons_founder</span>,</span>
<span>  <span class="st">"B.FR.83.HXB2_LAI_IIIB_BRU.K03455"</span>,</span>
<span>  start <span class="op">=</span> <span class="va">gp120_start</span>, end <span class="op">=</span> <span class="va">gp120_end</span></span>
<span><span class="op">)</span><span class="op">$</span><span class="va">founder</span><span class="op">)</span></span>
<span><span class="va">flt_gp120_end</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cumsum.html" class="external-link">cumsum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">hiv_env_flt_2022</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span><span class="op">)</span> <span class="op">!=</span> <span class="st">"-"</span><span class="op">)</span> <span class="op">==</span></span>
<span>  <span class="va">len_hxb2_gp120</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co"># subset to only gp120 section</span></span>
<span><span class="va">hiv_gp120_flt_2022</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slice_aln.html">slice_aln</a></span><span class="op">(</span><span class="va">hiv_env_flt_2022</span>, <span class="fl">1</span>, <span class="va">flt_gp120_end</span><span class="op">)</span></span></code></pre></div>
<p>Then we can use this sliced alignment to identify conserved
sites:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/identify_conserved_sites.html">identify_conserved_sites</a></span><span class="op">(</span><span class="va">hiv_gp120_flt_2022</span>, <span class="st">"B.FR.83.HXB2_LAI_IIIB_BRU.K03455"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,533 × 5</span></span></span>
<span><span class="co">#&gt;    founder_pos founder_base consensus_base consensus_prop conserved</span></span>
<span><span class="co">#&gt;          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>           0 a            a                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>           1 t            t                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>           2 g            g                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>           3 a            a                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>           4 g            g                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>           5 a            a                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>           6 g            g                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>           7 t            t                         0.8 No       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>           8 g            g                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>           9 a            a                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1,523 more rows</span></span></span></code></pre></div>
<p>Alternatively, if you have two alignments with a shared reference,
one from which you’d like to calculate conserved sites, and the other
that contains the founder, you can provide both alignments, as well as
the common reference, and the function will return the conserved sites
relative to the founder sequence. In this case, the shared reference
sequence is assumed to have the same start position in each
alignment.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">founder_conserved_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/identify_conserved_sites.html">identify_conserved_sites</a></span><span class="op">(</span><span class="va">hiv_gp120_flt_2022</span>,</span>
<span>  founder <span class="op">=</span> <span class="st">"B.US.2011.DEMB11US006.KC473833"</span>,</span>
<span>  ref <span class="op">=</span> <span class="st">"B.FR.83.HXB2_LAI_IIIB_BRU.K03455"</span>,</span>
<span>  founder_aln <span class="op">=</span> <span class="va">gp120</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,503 × 5</span></span></span>
<span><span class="co">#&gt;    founder_pos founder_base consensus_base consensus_prop conserved</span></span>
<span><span class="co">#&gt;          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>           0 a            a                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>           1 t            t                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>           2 g            g                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>           3 a            a                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>           4 g            g                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>           5 a            a                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>           6 g            g                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>           7 c            t                         0.8 No       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>           8 g            g                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>           9 a            a                         1   Yes      </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1,493 more rows</span></span></span></code></pre></div>
<p>You can visualize the conserved sites across the genome using the
following code:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">founder_conserved_df</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">founder_pos</span>, y <span class="op">=</span> <span class="va">consensus_prop</span>, fill <span class="op">=</span> <span class="va">conserved</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 30 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_col()`).</span></span></code></pre></div>
<p><img src="prepare_input_data_files/figure-html/viz_conserved-1.png" width="700"></p>
<p>The blue sites are conserved, the red sites are not conserved, and
the y-axis indicates the proportion of sequences in the alignment that
contain the consensus base. <strong>Note that this example only uses 10
sequences. We recommend using a large alignment with very diverse
sequences to identify conserved sites to ensure that the identified
sites are across many genetic backgrounds.</strong></p>
<p>While we provide information about each sequence in the output, the
input of <code>run_wavess</code> only takes a vector of the conserved
sites. To generate this, you can run the following code:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">founder_conserved_sites_example</span> <span class="op">&lt;-</span> <span class="va">founder_conserved_df</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">conserved</span> <span class="op">==</span> <span class="st">"Yes"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">founder_pos</span>, <span class="va">founder_base</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/enframe.html" class="external-link">deframe</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">toupper</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">founder_conserved_sites_example</span><span class="op">)</span></span>
<span><span class="co">#&gt;   0   1   2   3   4   5 </span></span>
<span><span class="co">#&gt; "A" "T" "G" "A" "G" "A"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">founder_conserved_sites_example</span><span class="op">)</span></span>
<span><span class="co">#&gt; 1497 1498 1499 1500 1501 1502 </span></span>
<span><span class="co">#&gt;  "A"  "A"  "A"  "A"  "G"  "A"</span></span></code></pre></div>
<p>To write this to a file (note that we use the internal conserved
sites because it is more accurate for our example, rather than the
example run above, which is not accurate since it was only run on a
handful of genomes):</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_csv</a></span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/enframe.html" class="external-link">enframe</a></span><span class="op">(</span><span class="va">founder_conserved_sites</span>, name <span class="op">=</span> <span class="st">"position"</span>, value <span class="op">=</span> <span class="st">"nucleotide"</span><span class="op">)</span>, <span class="st">"input_data/founder_conserved_sites.csv"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="find-a-consensus-sequence">Find a consensus sequence<a class="anchor" aria-label="anchor" href="#find-a-consensus-sequence"></a>
</h3>
<p>We have implemented a very crude way of generating a consensus
sequence, by taking the most common base at each position, and if there
is a tie, the base that comes first in the alphabet. The consensus
sequence is returned as part of the
<code><a href="../reference/identify_conserved_sites.html">identify_conserved_sites()</a></code> output. If you would like more
control over generating a consensus sequence, you can use the <a href="https://www.hiv.lanl.gov/content/sequence/CONSENSUS/consensus.html" class="external-link">Consensus
Maker</a> tool on the LANL HIV website.</p>
<p>To convert the consensus sequence from above into the correct input
format for <code><a href="../reference/run_wavess.html">run_wavess()</a></code>, you can use the following code,
where NA values are converted to gaps:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"NA"</span>, <span class="st">"-"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">founder_conserved_df</span><span class="op">$</span><span class="va">consensus_base</span>, collapse <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "atgagagtgatgg---------ggacacagatg---aagtggagatgggggactatgatcttgggaatgataataatttgtagtgctacagaaaacttgtgggttactgtctactatggggtacctgtgtggaaagatgcagagaccaccctattttgtgcatcagatgctaaagcatatgatacagaagtgcataatgtctgggctacacatgcctgtgtacccacagaccccaacccacaagaaataaatttggaaaatgtgacagaagagtttaacatgtggaaaaataacatggtagaacagatgcataaagatataatcagtctatgggaccaaagcctaaagccatgtgtaaagttaacccctctctgtgttactttaaagtgcaatgactacaacaacaaca--aa---cactactacaactgaggaaggagaaataaaaaactgctctttcaatatgaccacagaattaagagataagaaacagaaagtatattcacttttttatagacttgatatagtacaaattgataataa---------t------aagtatagattaataaattgtaatacctcagccattacacaggcttgtccaaaggtatcctttgagccaattcccatacattattgtgccccagctggttttgcgattctaaagtgtaatgataaggagttcaatggaacaggaccatgcaagaatgtcagcacagtacaatgcacacatggaatcaagccagtagtatcaactcaactgctgttaaatggcagtctagcagaagaagagataatgattagatctgaaaatatcacagacaatgccaaaaccataatagtacaacttaataagcctgtaaaaattaattgtaccagacctaacaacaatacaagaaaaagtatacat--aggaccaggacaagcattctatgcaacaggtga---cataggggatataagaaaagcacattgtaatgtcagtagaacagaatggaataaaactttacaaaaggtagccaaacaattaagaaaacactttaacaaaacaataatcttt---aataattcaggaggggatttagaaattacaacacatagttttaattgtggaggagaatttttctactgcaacacatcaggcctgtttaatagcacttggaataaaaacaataacacaaaaaaaaataaaactataactctcccatgcagaataaagcaaattataaatatgtggcagagagcaggacaagcaatatatgcccctcccatccaaggagtaataaggtgtgaatcaaacattacaggactactattaacaagagatggtggaaa------taataaa------aataccgaaaccttcagacctggaggaggagatatgagggacaattggagaagtgaattatataaatataaagtagtaaaaattgaaccactaggagtagcacccaccaaggcaaaaagaagagtggtggagagagaaaaaaga"</span></span></code></pre></div>
<p>If you have an alignment including the founder sequence and wavess
reference sequence you’d like to use, then you can also use the
<code><a href="../reference/extract_seqs.html">extract_seqs()</a></code> function to obtain both at once:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">founder_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_seqs.html">extract_seqs</a></span><span class="op">(</span><span class="va">hxb2_cons_founder</span>,</span>
<span>  founder <span class="op">=</span> <span class="st">"B.US.2011.DEMB11US006.KC473833"</span>,</span>
<span>  ref <span class="op">=</span> <span class="st">"CON_B(1295)"</span>,</span>
<span>  start <span class="op">=</span> <span class="va">gp120_start</span>, end <span class="op">=</span> <span class="va">gp120_end</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; $founder</span></span>
<span><span class="co">#&gt; [1] "ATGAGAGCGATGGGGATCATGAGGAATTGGCAACACTTGTGGAGATGGGGCATGATGCTCCTTGGGATGTTGATGATCTGTAATGCTACAGACAACTTGTGGGTCACAGTCTATTATGGGGTACCTGTGTGGAGGGAAGCAAACACAACTCTATTTTGTGCATCAGATGCTAAAGCATATGAGACAGAGGTACATAATGTTTGGGCCACACATGCCTGTGTACCCACAGACCCCAACCCACAAGAAGTAAAATTGGGAAATGTGACAGAAAATTTTAATGCATGGAAAAATGACATGGTAGAACAGATGCATGAGGATATAATCAGTCTATGGGATCAAAGCCTAAAGCCATGTGTAAGATTAACCCCACTCTGTGTTACTCTAAATTGCACTGATCTTAATGCCACTAGCATTGGTAGTAACATGACACTGAAGGGAGAAATAAAAAATTGCACTTTCAATATCACCACAAGTAAAAACGATAAAAAGACAACAGAACGTGCATATTTTAATAGACTTGATGTGGTACCAATGGATGATAATAGTAGTAGTAGTACTAGTTATAGGTTGATAAGTTGTAACACCTCAGTCATTACACATGCCTGCCCAAAGGTATCCTTTGAGCCAATTCCCATACATTATTGTGCCCCAGCTGGTTTTGCGATTCTAAAGTGTAATGATAAAAAATTTAATGGAAAAGGACTATGTAAAAATGTCAGCACAGTACAATGTACACATGGAATTAGACCAGTAGTATCAACTCAACTGTTGCTGAATGGCAGTCTAGCAGAAGAAGAAGTAGTAATTAGATCTGAAAATATCTCTAACAATGCCAAAACCATAATAGTACATCTGAAGGAATCTGTACAAATTATTTGTGTAAGACCCAACAACAATACAAGACAAGGTATACATATGGGACCAGGAAGGACATTTTATACAACAGGGGGGATAATAGGAGATATAAGGCAAGCATATTGTAACATTAGTAGGGCAGAATGGACTAACACTCTAGGAAAGATAGTTGGAAAATTAAGAGAACGATTTAATAAAACAATAATCTTTAATCATTCCTCAGGAGGGGACCTAGAAATTGTGACACACAGTTTTAATTGTGGAGGGGAATTTTTCTACTGCAATACATCAGCACTGTTTAATAGTACTTGGAATAGTACTATAAATACAAGTGAAAATGACACAATCATACTCCCATGCAGAATAAAACAAATTATAAATCTGTGGCAGGAAGTAGGAAGAGCAATGTATGCTCCTCCCATCAGGGGAAACATTAGCTGTACATCAAATATTACGGGGGTGCTATTAACAAGAGATGGTGGCGATGACCCTAACGGGACCAACGACACCGAGACCTTCAGACCTGGAGGAGGAGATATGAGGGACAATTGGAGAAATGAATTGTATAAATACAAAGTAGTAAAAATTGAACCATTGGGAATAGCACCCACCAGGGCAAAGAGAAGAGTGGTGCAAAGAGAAAAAAGA"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $ref</span></span>
<span><span class="co">#&gt; [1] "ATGAGAGTGAAGGGGATCAGGAAGAATTATCAGCACTTGTGGAGATGGGGCATCATGCTCCTTGGGATGTTGATGATCTGTAGTGCTGCAGAAAAATTGTGGGTCACAGTCTATTATGGGGTACCTGTGTGGAAAGAAGCAACCACCACTCTATTTTGTGCATCAGATGCTAAAGCATATGATACAGAGGTACATAATGTTTGGGCCACACATGCCTGTGTACCCACAGACCCCAACCCACAAGAAGTAGTATTGGAAAATGTGACAGAAAATTTTAACATGTGGAAAAATAACATGGTAGAACAGATGCATGAGGATATAATCAGTTTATGGGATCAAAGCCTAAAGCCATGTGTAAAATTAACCCCACTCTGTGTTACTTTAAATTGCACTGATTTTAATACTAATAATAATAATACTAATA-TA-TATGAAAGGAGAAATAAAAAACTGCTCTTTCAATATCACCACAAGCATAAGAGATAAGATGCAGAAAGAATATGCACTTTTTTATAAACTTGATGTAGTACCAATAGATAATGA---------TAATACTAGCTATAGGTTGATAAGTTGTAACACCTCAGTCATTACACAGGCCTGTCCAAAGGTATCCTTTGAGCCAATTCCCATACATTATTGTGCCCCGGCTGGTTTTGCGATTCTAAAGTGTAATGATAAGAAGTTCAATGGAACAGGACCATGTAAAAATGTCAGCACAGTACAATGTACACATGGAATTAGGCCAGTAGTATCAACTCAACTGCTGTTAAATGGCAGTCTAGCAGAAGAAGAGGTAGTAATTAGATCTGAAAATTTCACAGACAATGCTAAAACCATAATAGTACAGCTGAATGAATCTGTAGAAATTAATTGTACAAGACCCAACAACAATACAAGAAAAAGTATACATATAGGACCAGGGAGAGCATTTTATGCAACAGGAGAAATAATAGGAGATATAAGACAAGCACATTGTAACATTAGTAGAGCAAAATGGAATAACACTTTAAAACAGATAGTTAAAAAATTAAGAGAACAATTTAATAAAACAATAGTCTTTAATCAATCCTCAGGAGGGGACCCAGAAATTGTAATGCACAGTTTTAATTGTGGAGGGGAATTTTTCTACTGTAATACAACACAACTGTTTAATAGTACTTGGAATAATAATAATA-TACTAAAAAAAATGAAACTATCACACTCCCATGCAGAATAAAACAAATTATAAACATGTGGCAGGAAGTAGGAAAAGCAATGTATGCCCCTCCCATCAGAGGACAAATTAGATGTTCATCAAATATTACAGGGCTGCTATTAACAAGAGATGGTGGTAA------TAATAATA--AACAACACTGAGACCTTCAGACCTGGAGGAGGAGATATGAGGGACAATTGGAGAAGTGAATTATATAAATATAAAGTAGTAAAAATTGAACCATTAGGAGTAGCACCCACCAAGGCAAAGAGAAGAGTGGTGCAGAGAGAAAAAAGA"</span></span></code></pre></div>
<p>It’s okay if the reference sequence has gaps. These will be ignored
when computing fitness relative to the reference.</p>
<p>To write this to a fasta file, you use a similar method as writing
the founder to a fasta file:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">founder_ref</span><span class="op">$</span><span class="va">ref</span>, <span class="st">""</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ape/man/as.alignment.html" class="external-link">as.DNAbin</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ape/man/write.dna.html" class="external-link">write.FASTA</a></span><span class="op">(</span><span class="st">"input_data/ref.fasta"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="sample-epitopes">Sample epitopes<a class="anchor" aria-label="anchor" href="#sample-epitopes"></a>
</h3>
<p>Immune fitness in <code>wavess</code> is defined by epitope locations
in the sequence, each of which can have a maximum fitness cost between 0
and 1. These are defined at the amino acid level, so they must both be
within a nucleotide sequence that is translated to a protein and in the
correct reading frame. Note that, currently, epitope locations must all
be the same length.</p>
<p>While you can define your own epitope locations, we provide the
function <code><a href="../reference/sample_epitopes.html">sample_epitopes()</a></code> that, given epitope
probabilities for each <em>amino acid</em> position of interest, will
return randomly sampled <em>nucleoitde</em> epitope locations based on
the probability of an epitope occurring at that location. Therefore, the
function <code><a href="../reference/sample_epitopes.html">sample_epitopes()</a></code> takes as input a data frame that
must contain columns for the amino acid position
(<code>aa_position</code>) and the epitope probability at that position
(<code>epitope_probability</code>). The amino acid positions must only
include those corresponding to the nucleotide sequence to be simulated,
and must be indexed in such a way that the first amino acid position
corresponds to the first nucleotide position in the founder sequence. In
our example, this is gp120. We have already filtered the built-in
package data to only include this subset.</p>
<p>If possible, we recommend determining epitope probabilities based on
some sort of known antibody contact/binding/neutralization maps. For HIV
ENV gp120, we used the features from the LANL HIV immunology
database:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">env_features</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3,022 × 13</span></span></span>
<span><span class="co">#&gt;       ID Title                   Reference `Feature type` `Mab(s)(Binding type)`</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     4 PGT121 epitope defined… Mouquet2… binding        PGT121(V3)            </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     4 PGT121 epitope defined… Mouquet2… binding        PGT121(V3)            </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     4 PGT121 epitope defined… Mouquet2… binding        PGT121(V3)            </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>     4 PGT121 epitope defined… Mouquet2… binding        PGT121(V3)            </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>     4 PGT121 epitope defined… Mouquet2… binding        PGT121(V3)            </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>     4 PGT121 epitope defined… Mouquet2… binding        PGT121(V3)            </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>     4 PGT121 epitope defined… Mouquet2… binding        PGT121(V3)            </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>     4 PGT121 epitope defined… Mouquet2… binding        PGT121(V3)            </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     4 PGT121 epitope defined… Mouquet2… binding        PGT121(V3)            </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>     4 PGT121 epitope defined… Mouquet2… binding        PGT121(V3)            </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3,012 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 8 more variables: `Experimental method(s)` &lt;chr&gt;, Position &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   `Env feature(s)` &lt;chr&gt;, `HXB2 AA` &lt;chr&gt;, `Entropy M` &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   `Entropy B` &lt;dbl&gt;, `Entropy C` &lt;dbl&gt;, Annotation &lt;chr&gt;</span></span></span></code></pre></div>
<p>All we need is the positions column, but we provide the rest of the
information for reference.</p>
<p>If you don’t have this information, you can use uniformly distributed
probabilities, or some other informed guess as to where the
probabilities for epitopes binding is higher.</p>
<p>We can put the positions into the
<code><a href="../reference/get_epitope_frequencies.html">get_epitope_frequencies()</a></code> function, which returns a tibble
with three columns:</p>
<ul>
<li>
<code>aa_pos</code>: amino acid position</li>
<li>
<code>n_features</code>: the number of features at that
position</li>
<li>
<code>epitope_probability</code>: the probability of an epitope at
that position, given the input positions</li>
</ul>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">epi_probs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_epitope_frequencies.html">get_epitope_frequencies</a></span><span class="op">(</span><span class="va">env_features</span><span class="op">$</span><span class="va">Position</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 480 × 3</span></span></span>
<span><span class="co">#&gt;    aa_position n_features epitope_probability</span></span>
<span><span class="co">#&gt;          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>          31          1            0.000<span style="text-decoration: underline;">331</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>          32          0            0       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>          33          0            0       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>          34          0            0       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>          35          0            0       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>          36          0            0       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>          37          0            0       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>          38          0            0       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>          39          0            0       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>          40          0            0       </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 470 more rows</span></span></span></code></pre></div>
<p>We can visualize it as follows:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">epi_probs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">aa_position</span>, y <span class="op">=</span> <span class="va">epitope_probability</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"HXB2 amino acid position"</span>, y <span class="op">=</span> <span class="st">"Epitope probability"</span><span class="op">)</span></span></code></pre></div>
<p><img src="prepare_input_data_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>This information can be used as input to the
<code><a href="../reference/sample_epitopes.html">sample_epitopes()</a></code> function. Note that the output of this
function is different each time it’s run, since the locations are
selected randomly each time.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sample_epitopes.html">sample_epitopes</a></span><span class="op">(</span><span class="va">epi_probs</span><span class="op">)</span></span>
<span><span class="co">#&gt; 2 resamples required</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 10 × 3</span></span></span>
<span><span class="co">#&gt;    epi_start_nt epi_end_nt max_fitness_cost</span></span>
<span><span class="co">#&gt;           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>          483        513             0.03</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>          906        936             0.06</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>          981       <span style="text-decoration: underline;">1</span>011             0.09</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>          822        852             0.12</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>         <span style="text-decoration: underline;">1</span>083       <span style="text-decoration: underline;">1</span>113             0.15</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>         <span style="text-decoration: underline;">1</span>488       <span style="text-decoration: underline;">1</span>518             0.18</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>          525        555             0.21</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>          405        435             0.24</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>          441        471             0.27</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>         <span style="text-decoration: underline;">1</span>368       <span style="text-decoration: underline;">1</span>398             0.3</span></span></code></pre></div>
<p>The function returns start and end <em>nucleotide</em> positions for
each epitope, relative to the input amino acid positions.</p>
<p>There are many ways to customize the output of this function. You can
provide the starting (<code>start_aa_pos</code>) and ending
(<code>end_aa_pos</code>) amino acid positions to consider for epitope
sampling (here we’ve pre-subset to only gp120 features), the number of
epitopes to sample (<code>num_epitopes</code>), the amino acid epitope
length (<code>aa_epitope_length</code>), and the maximum fitness cost of
an epitope (<code>max_fit_cost</code>).</p>
<p>If the amino acid positions are relative to a community reference
sequence, then you will also need to create a map between the community
reference and your founder sequence so that the data can be re-indexed
to the founder sequence. In this case, the amino acid positions we used
are from HXB2, so we want to map them to our founder sequence. You
should use an alignment that contains the exact founder sequence you
plan to simulate (no longer and no shorter). Once you have this
alignment (e.g. using the <code><a href="../reference/slice_aln.html">slice_aln()</a></code> function described
above), you can use the <code><a href="../reference/map_ref_founder.html">map_ref_founder()</a></code> function, which
returns reference and founder positions mapped to each other:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ref_founder_map</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/map_ref_founder.html">map_ref_founder</a></span><span class="op">(</span><span class="va">gp120</span>,</span>
<span>  ref <span class="op">=</span> <span class="st">"B.FR.83.HXB2_LAI_IIIB_BRU.K03455"</span>,</span>
<span>  founder <span class="op">=</span> <span class="st">"B.US.2011.DEMB11US006.KC473833"</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,563 × 5</span></span></span>
<span><span class="co">#&gt;    alignment_pos ref_pos founder_pos ref_base founder_base</span></span>
<span><span class="co">#&gt;            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>             0       0           0 a        a           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>             1       1           1 t        t           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>             2       2           2 g        g           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>             3       3           3 a        a           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>             4       4           4 g        g           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>             5       5           5 a        a           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>             6       6           6 g        g           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>             7       7           7 t        c           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>             8       8           8 g        g           </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>             9       9           9 a        a           </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1,553 more rows</span></span></span></code></pre></div>
<p>You can input this into <code><a href="../reference/sample_epitopes.html">sample_epitopes()</a></code>, which will
then return nucleotide positions relative to the founder sequence
(instead of HXB2):</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">epitope_locations</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sample_epitopes.html">sample_epitopes</a></span><span class="op">(</span><span class="va">epi_probs</span>,</span>
<span>  ref_founder_map <span class="op">=</span> <span class="va">ref_founder_map</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; 4 resamples required</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 10 × 3</span></span></span>
<span><span class="co">#&gt;    epi_start_nt epi_end_nt max_fitness_cost</span></span>
<span><span class="co">#&gt;           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>         <span style="text-decoration: underline;">1</span>053       <span style="text-decoration: underline;">1</span>083             0.03</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>          930        960             0.06</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>          582        612             0.09</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>         <span style="text-decoration: underline;">1</span>392       <span style="text-decoration: underline;">1</span>422             0.12</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>         <span style="text-decoration: underline;">1</span>245       <span style="text-decoration: underline;">1</span>275             0.15</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>          435        465             0.18</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>          969        999             0.21</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>          522        552             0.24</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>          816        846             0.27</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>          474        504             0.3</span></span></code></pre></div>
<p>Here are two ways to visualize the epitopes:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">epitope_locations</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">epi_start_nt</span>, y <span class="op">=</span> <span class="va">max_fitness_cost</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Epitope start position"</span>, y <span class="op">=</span> <span class="st">"Max fitness cost"</span><span class="op">)</span></span></code></pre></div>
<p><img src="prepare_input_data_files/figure-html/plot_epi-1.png" width="700"></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">epitope_locations</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">max_fitness_cost</span>, col <span class="op">=</span> <span class="va">epi_start_nt</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Epitope"</span>, y <span class="op">=</span> <span class="st">"Max fitness cost"</span>, col <span class="op">=</span> <span class="st">"Epitope start\nposition"</span><span class="op">)</span></span></code></pre></div>
<p><img src="prepare_input_data_files/figure-html/plot_epi-2.png" width="700"></p>
<p>To write them to a file:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_csv</a></span><span class="op">(</span><span class="va">epitope_locations</span>, <span class="st">"input_data/epitope_locations.csv"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="on-to-simulations">On to simulations<a class="anchor" aria-label="anchor" href="#on-to-simulations"></a>
</h2>
<p>Next, check out <code><a href="../articles/run_wavess.html">vignette("run_wavess")</a></code> to see how to use
all these inputs to simulate within-host evolution.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Zena Lapp, Narmada Sambaturu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
