<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Running wavess in Python • wavess</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Running wavess in Python">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">wavess</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/prepare_input_data.html">Prepare input data</a></li>
    <li><a class="dropdown-item" href="../articles/run_wavess.html">Simulate within-host evolution with `wavess`</a></li>
    <li><a class="dropdown-item" href="../articles/analyze_output.html">Analyze wavess output</a></li>
    <li><a class="dropdown-item" href="../articles/python.html">Running wavess in Python</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Running wavess in Python</h1>
            
      

      <div class="d-none name"><code>python.Rmd</code></div>
    </div>

    
    
<p>We provide the R function <code><a href="../reference/run_wavess.html">run_wavess()</a></code> as a convenience
function to run the simulator from R, but since the underlying functions
are all written in Python, you can also run wavess with Python using a
command-line script. We provide an example Python script
<code>run_wavess.py</code> (in <code>inst/python/</code>) to make this
easy to do. The script takes a config file
(<code>inst/python/config.yaml</code>) that includes paths to input
files and defines various input arguments such as what selective
pressures to model, as well as an output file path/prefix.</p>
<p><strong>Please note that the python script does not check the inputs
as comprehensively as the R function. If you need to troubleshoot an
error you are getting with the Python script, it is probably worth
testing it out in the R function to see if you get a more informative
error message.</strong></p>
<p>We provide example input files in <code>inst/extdata</code>. You can
generate the input files on your own, or using the functions provided in
this package. See <code><a href="../articles/prepare_input_data.html">vignette("prepare_input_data")</a></code> for
details on how to create and save the required input files. Here, we
briefly outline what goes in the config file and how to run
<code>run_wavess.py</code> itself.</p>
<div class="section level3">
<h3 id="installing-dependencies">Installing dependencies<a class="anchor" aria-label="anchor" href="#installing-dependencies"></a>
</h3>
<p>The Python script implementation of wavess requires the following
imports: pandas, Bio, time, sys, os, yaml, csv, copy, random, numpy,
collections</p>
<p>The easiest way to install these is to create a conda environment. f
you don’t already have it installed, first install <a href="https://mamba.readthedocs.io/en/latest/installation/mamba-installation.html" class="external-link">mamba</a>
(or conda, but it’s slower).</p>
<p>Then, create the wavess conda environment (you only have to do this
once), you can use the environment yaml file in the
<code>inst/python</code> directory. Navigate to that directory and then
run the command:</p>
<pre><code>mamba env create -f env.yaml</code></pre>
<p>Next, activate the wavess environment (you have to do this once per
session):</p>
<pre><code>mamba activate wavess</code></pre>
</div>
<div class="section level3">
<h3 id="config-file">Config file<a class="anchor" aria-label="anchor" href="#config-file"></a>
</h3>
<p><code>run_wavess.py</code> requires a <code>config.yaml</code> file
(example config file in <code>inst/python</code>) that defines various
parameters and includes paths to input files.</p>
<p><strong>File inputs (example data in
<code>inst/extdata/</code>):</strong></p>
<ul>
<li>
<code>pop_samp</code>: csv with columns generation,
active_cell_count, n_sample_active. Note that the initial active cell
population size in the first generation must be the same as the number
of input founder sequences (because it simply <em>is</em> the input
founder sequences). Can be generated using the [generate_pop_samp()]
function.</li>
<li>
<code>founder_seqs</code>: fasta file of founder sequence(s). The
founder sequence(s) may only contain the characters ACGT, and no gaps
are allowed. When modeling immune fitness, they are expected to be
codon-aligned.</li>
<li>
<code>q</code>: csv that is a named Q matrix of nucleotide
substitution rates. Rows are from, columns are to. Can be generated
using the [estimate_q()] function.</li>
<li>
<code>conserved_sites</code>: csv of conserved sites indexed at 0
where each row contains a conserved site in a column called “position”
and the conserved nucleotide found at that site in a column called
“nucleotide”. This can be generated using the
[identify_conserved_sites()] function. If this is an empty string, then
conserved fitness costs are not included.</li>
<li>
<code>ref_seq</code>: fasta file of the reference sequence. A
consensus sequence, that can be used as the reference sequence, can be
generated using the function [identify_conserved_sites()]. If this is an
empty string, then fitness costs relative to a reference sequence are
not included.</li>
<li>
<code>epitope_locations</code>: csv of epitope locations and maximum
fitness costs with columns epi_start_nt, epi_end_nt, max_fitness_cost.
These epitopes are expected to be indexed at 0 and in a protein in the
correct reading frame, as the nucleotide sequences are translated to
amino acids to calculate the immune fitness cost. This information can
be generated using the functions [get_epitope_frequencies()] and
[sample_epitopes()]. If this is an empty string provided, then immune
fitness costs are not included.</li>
</ul>
<p><strong>Parameters:</strong></p>
<p>Diversity-generating mechanisms:</p>
<ul>
<li>
<code>mut_rate</code>: Mutation rate per site per generation</li>
<li>`recomb_rate``: Recombination rate per site per generation</li>
</ul>
<p>Latent cell dynamics:</p>
<ul>
<li>
<code>prob_act_to_lat</code>: probability of infected cell becoming
latent</li>
<li>
<code>prob_lat_to_act</code>: probability that a latent cell will
become active</li>
<li>
<code>prob_lat_prolif</code>: probability that a latent cell will
proliferate (based on central memory cells)</li>
<li>
<code>prob_lat_die</code>: probability that a latent cell will
die</li>
</ul>
<p>Parameters related to fitness:</p>
<p>Conserved sites:</p>
<ul>
<li>
<code>conserved_cost</code>: fitness cost per mutation in a
conserved site</li>
</ul>
<p>Fitness relative to a reference sequence:</p>
<ul>
<li>
<code>rep_exp</code>: exponent for replicative fitness function</li>
</ul>
<p>Immune fitness:</p>
<ul>
<li>
<code>seroconversion_time</code>: seroconversion time in
generations; immunity considered only at this generation onward</li>
<li>
<code>immune_response_proportion</code>: proportion new viral
antigen must reach to initiate immune response against corresponding
epitope</li>
<li>
<code>time_to_full_potency</code>: number of generations for a new
immune response to reach full potency</li>
</ul>
</div>
<div class="section level3">
<h3 id="running-run_wavess-py">Running <code>run_wavess.py</code><a class="anchor" aria-label="anchor" href="#running-run_wavess-py"></a>
</h3>
<p><code>run_wavess.py</code> takes two required inputs and one optional
input:</p>
<ul>
<li>The config file, which contains information about input file paths
and parameters</li>
<li>The prefix to the output files (including directories)</li>
<li>An optional number that is the seed to set for random number
generation (to make your results reproducible)</li>
</ul>
<p>Here is an example (we set a seed so it will be reproducible, usually
you don’t want to do this):</p>
<pre><code>python run_wavess.py config.yaml wavess_output/ 1234</code></pre>
<p><strong>Note that the path to the output file has to have at least
one forward slash.</strong> If the slash is at the end, then there will
be no prefix for the output files. If you want a prefix for the output
files, then you can add that to the end.</p>
<p>Two output files are generated: a csv file containing various counts
for each sampled generation, and a fasta file containing the sampled
sequences for each generation. See <code><a href="../articles/run_wavess.html">vignette("run_wavess")</a></code>
for more details about the outputs.</p>
<p>Example output files can be found in
<code>inst/python/wavess_output/</code>.</p>
</div>
<div class="section level3">
<h3 id="python-tests">Python tests<a class="anchor" aria-label="anchor" href="#python-tests"></a>
</h3>
<p>To run the tests for the Python part of the code, navigate to
<code>inst/python</code> and run:</p>
<pre><code>pytest tests.py --cov --cov-report=html</code></pre>
<p>You can then look at the coverage report in
<code>htmlcov/index.html</code>.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Zena Lapp, Narmada Sambaturu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
