<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Simulate within-host evolution with `wavess` • wavess</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate within-host evolution with `wavess`">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">wavess</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/prepare_input_data.html">Prepare input data</a></li>
    <li><a class="dropdown-item" href="../articles/run_wavess.html">Simulate within-host evolution with `wavess`</a></li>
    <li><a class="dropdown-item" href="../articles/analyze_output.html">Analyze wavess output</a></li>
    <li><a class="dropdown-item" href="../articles/python.html">Running wavess in Python</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Simulate within-host evolution with `wavess`</h1>
            
      

      <div class="d-none name"><code>run_wavess.Rmd</code></div>
    </div>

    
    
<p>wavess is a agent-based discrete-time within-host evolution
simulator. We assume that all virus life cycles within the infection are
synchronized into generations. Each generation is one full virus life
cycle, from infecting a cell to exiting the cell.</p>
<p>In this vignette, we focus on how to simulate within-host evolution
using the <code><a href="../reference/run_wavess.html">run_wavess()</a></code> function. We will be using the HIV
ENV gp120 gene as an example. Please note that the default arguments
were set with this particular organism and genomic region in mind. If
you’d like to simulate something else, you may have to modify certain
parameters. However, if you are interested in this gene in particular,
you can probably use most of the defaults including the founder and
reference sequences provided as examples.</p>
<div class="section level2">
<h2 id="a-note-about-using-the-simulation-output">A note about using the simulation output<a class="anchor" aria-label="anchor" href="#a-note-about-using-the-simulation-output"></a>
</h2>
<p>Before we get started, I just want to note that the output of these
simulations is stochastic, meaning they will be different each time it
is run. You can set a seed for reproducibility and troubleshooting if
you’d like, but for the vast majority of applications, you will want to
run many replicates of a given simulation scenario with different seeds
(random is okay) to get a reasonable sense of what dynamics are
observed. You will also want to check the simulation outputs to see if
they are giving you reasonable/expected results that conform to your
prior knowledge about the system you’re simulating. If the output does
not seem right, you might have to tweak the input parameters to work for
your particular system. We go into more detail about analyzing the model
output in a separate vignette (see
<code><a href="../articles/analyze_output.html">vignette("analyze_output")</a></code>).</p>
</div>
<div class="section level2">
<h2 id="load-libraries">Load libraries<a class="anchor" aria-label="anchor" href="#load-libraries"></a>
</h2>
<p>First, we have to load the wavess library and a few additional
libraries that we’ll use for this vignette:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://molevolepid.github.io/wavess/">wavess</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/emmanuelparadis/ape" class="external-link">ape</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:ape':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     where</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-python-virtual-environment">Create Python virtual environment<a class="anchor" aria-label="anchor" href="#create-python-virtual-environment"></a>
</h2>
<p>Next, we must create a Python virtual environment. The virtual
environment is required because the underlying code of
<code><a href="../reference/run_wavess.html">run_wavess()</a></code> is written in Python. You only have to create
the virtual environment once on each machine.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/create_python_venv.html">create_python_venv</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in create_python_venv(): Skipped installation of the following</span></span>
<span><span class="co">#&gt; packages: scipy Use `force` to force installation or update.</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="prepare-input-data">Prepare input data<a class="anchor" aria-label="anchor" href="#prepare-input-data"></a>
</h2>
<p>Next, we must generate the input data. If you’re interested in
learning more about how to prepare the input data, please see the
corresponding vignette (see
<code><a href="../articles/prepare_input_data.html">vignette("prepare_input_data")</a></code>).</p>
<p>We will only simulate 300 generations with sampling every 100
generations so the simulation doesn’t take too long to run.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pop</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/define_growth_curve.html">define_growth_curve</a></span><span class="op">(</span>n_gens <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span>
<span><span class="va">samp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/define_sampling_scheme.html">define_sampling_scheme</a></span><span class="op">(</span>sampling_frequency_active <span class="op">=</span> <span class="fl">100</span>, sampling_frequency_latent <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">day</span> <span class="op">&lt;=</span> <span class="fl">300</span><span class="op">)</span></span>
<span><span class="va">founder_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_seqs.html">extract_seqs</a></span><span class="op">(</span><span class="va">hxb2_cons_founder</span>,</span>
<span>  founder <span class="op">=</span> <span class="st">"B.US.2011.DEMB11US006.KC473833"</span>,</span>
<span>  ref <span class="op">=</span> <span class="st">"CON_B(1295)"</span>,</span>
<span>  start <span class="op">=</span> <span class="fl">6225</span>, end <span class="op">=</span> <span class="fl">7787</span></span>
<span><span class="op">)</span></span>
<span><span class="va">gp120</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slice_aln.html">slice_aln</a></span><span class="op">(</span><span class="va">hxb2_cons_founder</span>, <span class="fl">6225</span>, <span class="fl">7787</span><span class="op">)</span></span>
<span><span class="va">epi_probs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_epitope_frequencies.html">get_epitope_frequencies</a></span><span class="op">(</span><span class="va">env_features</span><span class="op">$</span><span class="va">Position</span><span class="op">)</span></span>
<span><span class="va">ref_founder_map</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/map_ref_founder.html">map_ref_founder</a></span><span class="op">(</span><span class="va">gp120</span>,</span>
<span>  ref <span class="op">=</span> <span class="st">"B.FR.83.HXB2_LAI_IIIB_BRU.K03455"</span>,</span>
<span>  founder <span class="op">=</span> <span class="st">"B.US.2011.DEMB11US006.KC473833"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">epitope_locations</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sample_epitopes.html">sample_epitopes</a></span><span class="op">(</span><span class="va">epi_probs</span>,</span>
<span>  ref_founder_map <span class="op">=</span> <span class="va">ref_founder_map</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; 16 resamples required</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="simplest-way-to-run_wavess">Simplest way to <code>run_wavess()</code><a class="anchor" aria-label="anchor" href="#simplest-way-to-run_wavess"></a>
</h2>
<p>The simplest way to simulate within-host evolution is to use all of
the defaults, and only input the population growth and sampling scheme,
founder sequence, and nucleotide substitution probabilities.</p>
<p>Please note that the total population size simulated greatly
influences the simulation output.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">required_args_only</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_wavess.html">run_wavess</a></span><span class="op">(</span></span>
<span>  inf_pop_size <span class="op">=</span> <span class="va">pop</span>,</span>
<span>  samp_scheme <span class="op">=</span> <span class="va">samp</span>,</span>
<span>  founder_seqs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">founder_ref</span><span class="op">$</span><span class="va">founder</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Here, we simulate one founder sequence. You may simulate more, but
you will have to ensure that they are the same length, do not include
gaps, and are codon-aligned. You will also have to modify the input
growth curve to start with the correct number of cells. See
<code><a href="../articles/prepare_input_data.html">vignette('prepare_input_data')</a></code> for more details.</p>
<p>The output format of <code><a href="../reference/run_wavess.html">run_wavess()</a></code> is a list of length
four containing a tibble of various counts, the fitness of the sampled
viruses, and the DNA sequences of the sampled viruses from the active
and latent reservoir:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">required_args_only</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "counts"      "fitness"     "seqs_active" "seqs_latent"</span></span></code></pre></div>
<p>Note that, if no latent cells are sampled, then the length of the
list will be three instead, with no sequences from latent cells.</p>
<p>The counts tibble contains the following columns:</p>
<ul>
<li>
<code>generation</code>: generation that number of events was
recorded and sequences were sampled</li>
<li>
<code>active_cell_count</code>: active cell count</li>
<li>
<code>latent_cell_count</code>: latent cell count</li>
<li>
<code>active_turned_latent</code>: number of active cells that
became latent</li>
<li>
<code>latent_turned_active</code>: number of latent cells that
became active</li>
<li>
<code>latent_died</code>: number of latent cells that died</li>
<li>
<code>latent_proliferated</code>: number of latent cells that
proliferated</li>
<li>
<code>number_mutations</code>: number of mutations across all
sequences</li>
<li>
<code>number_recombinations</code>: number of infected cells with at
least one recombination event</li>
<li>
<code>mean_fitness_active</code>: mean fitness of active cells</li>
<li>
<code>mean_conserved_active</code>: mean conserved sites fitness of
active cells</li>
<li>
<code>mean_immune_active</code>: mean immune fitness of active
cells</li>
<li>
<code>mean_replicative_active</code>: mean replicative fitness
(comparison to <code>ref_seq</code>) of active cells</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">required_args_only</span><span class="op">$</span><span class="va">counts</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 13</span></span></span>
<span><span class="co">#&gt;   generation active_cell_count latent_cell_count active_turned_latent</span></span>
<span><span class="co">#&gt;        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>          0                10                 0                    0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>        100              <span style="text-decoration: underline;">2</span>000               116                    2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>        200              <span style="text-decoration: underline;">2</span>000               145                    2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>        300              <span style="text-decoration: underline;">2</span>000               178                    2</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 9 more variables: latent_turned_active &lt;int&gt;, latent_died &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   latent_proliferated &lt;int&gt;, number_mutations &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   number_recombinations &lt;int&gt;, mean_fitness_active &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   mean_conserved_active &lt;dbl&gt;, mean_immune_active &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   mean_replicative_active &lt;dbl&gt;</span></span></span></code></pre></div>
<p>As you can see, the default values for <code><a href="../reference/run_wavess.html">run_wavess()</a></code>
include latency, mutations, and dual infections (which lead to
recombination), but no fitness costs.</p>
<p>The fitness tibble contains a row for each sampled sequence (all from
the active pool) and the following columns:</p>
<ul>
<li>
<code>generation</code>: generation that the virus was sampled
in</li>
<li>
<code>seq_id</code>: sequence id (corresponds to the sequence name
below). Note that the same value for different generations <em>does
not</em> say anything about the ancestral relationship between the
viruses.</li>
<li>
<code>immune</code>: virus immune fitness</li>
<li>
<code>conserved</code>: virus conserved fitness</li>
<li>
<code>replicative</code>: virus replicative fitness (compared to a
reference)</li>
<li>
<code>overall</code>: overall fitness (the other three multiplied
together)</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">required_args_only</span><span class="op">$</span><span class="va">fitness</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 80 × 6</span></span></span>
<span><span class="co">#&gt;    generation seq_id   immune conserved replicative overall</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> founder    founder0      1         1           1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> founder    founder1      1         1           1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> founder    founder2      1         1           1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> founder    founder3      1         1           1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> founder    founder4      1         1           1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> founder    founder5      1         1           1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> founder    founder6      1         1           1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> founder    founder7      1         1           1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> founder    founder8      1         1           1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> founder    founder9      1         1           1       1</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 70 more rows</span></span></span></code></pre></div>
<p>Again, the fitness is 1 for everything here because we didn’t model
fitness.</p>
<p>The sequences are returned as well, in <code><a href="https://rdrr.io/pkg/ape/man/DNAbin.html" class="external-link">ape::DNAbin</a></code>
format:</p>
<p>Sequences from active cells:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">required_args_only</span><span class="op">$</span><span class="va">seqs_active</span></span>
<span><span class="co">#&gt; 80 DNA sequences in binary format stored in a matrix.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; All sequences of same length: 1503 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Labels:</span></span>
<span><span class="co">#&gt; founder0</span></span>
<span><span class="co">#&gt; founder1</span></span>
<span><span class="co">#&gt; founder2</span></span>
<span><span class="co">#&gt; founder3</span></span>
<span><span class="co">#&gt; founder4</span></span>
<span><span class="co">#&gt; founder5</span></span>
<span><span class="co">#&gt; ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Base composition:</span></span>
<span><span class="co">#&gt;     a     c     g     t </span></span>
<span><span class="co">#&gt; 0.370 0.165 0.222 0.243 </span></span>
<span><span class="co">#&gt; (Total: 120.24 kb)</span></span></code></pre></div>
<p>Sequences from latent cells:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">required_args_only</span><span class="op">$</span><span class="va">seqs_latent</span></span>
<span><span class="co">#&gt; 60 DNA sequences in binary format stored in a matrix.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; All sequences of same length: 1503 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Labels:</span></span>
<span><span class="co">#&gt; gen100_latent_0</span></span>
<span><span class="co">#&gt; gen100_latent_1</span></span>
<span><span class="co">#&gt; gen100_latent_2</span></span>
<span><span class="co">#&gt; gen100_latent_3</span></span>
<span><span class="co">#&gt; gen100_latent_4</span></span>
<span><span class="co">#&gt; gen100_latent_5</span></span>
<span><span class="co">#&gt; ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Base composition:</span></span>
<span><span class="co">#&gt;     a     c     g     t </span></span>
<span><span class="co">#&gt; 0.370 0.164 0.222 0.243 </span></span>
<span><span class="co">#&gt; (Total: 90.18 kb)</span></span></code></pre></div>
<p>Each sequence is named as follows:</p>
<ul>
<li>Founder sequences are named as “founderX” where X is the index of
the founder sequence in the input vector (indexed at 0).</li>
<li>All other sequences are sampled and are named by the generation,
whether they were sampled from the active or latent reservoir, and a
number. Note again that the same value for different generations
<em>does not</em> say anything about the ancestral relationship between
the viruses.</li>
</ul>
<p>These outputs can be plotted and analyzed in various ways. If you’d
like to learn more about how to analyze the output, please check out the
post-processing vignette <code><a href="../articles/analyze_output.html">vignette("analyze_output")</a></code>.</p>
</div>
<div class="section level2">
<h2 id="including-selective-pressures">Including selective pressures<a class="anchor" aria-label="anchor" href="#including-selective-pressures"></a>
</h2>
<p><code><a href="../reference/run_wavess.html">run_wavess()</a></code> can simulate three types of selective
pressure:</p>
<ul>
<li>Conserved sites fitness</li>
<li>Fitness relative to a “most fit” reference sequence</li>
<li>B-cell immune response</li>
</ul>
<p>Note that all nucleotide positions related to fitness are expected to
be indexed at 0. Including these requires additional inputs, which are
described in more detail in the prepare input data vignette
(<code><a href="../articles/prepare_input_data.html">vignette("prepare_input_data")</a></code>).</p>
<p>Both conserved sites fitness and replicative fitness (comparison to a
reference) are modeled using a multiplicative fitness landscape with the
equation:</p>
<pre><code><span><span class="va">fitness</span> <span class="op">=</span> <span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">cost</span><span class="op">)</span> <span class="op">**</span> <span class="va">n_mut</span></span></code></pre>
<p>The fitness cost must be between 0 and 1, where 0 indicates no
fitness cost and 1 indicates no ability to survive. <code>n_mut</code>
is the number of mutations in conserved sites or the number of
nucleotides different from the reference sequence.</p>
<p>For additional details on how each of these selective pressures is
implemented, please see the manuscript [<strong>ADD LINK WHEN WE HAVE
IT</strong>].</p>
<div class="section level3">
<h3 id="conserved-sites-fitness">Conserved sites fitness<a class="anchor" aria-label="anchor" href="#conserved-sites-fitness"></a>
</h3>
<p>To simulate conserved sites fitness, you must provide a vector of
sites in the founder sequence that are considered to be conserved
(<code>conserved_sites</code> argument). By default, the cost of a
mutation at these sites is 0.99 (<code>conserved_cost</code>
argument).</p>
<p>Here’s an example of simulating evolution with conserved sites
fitness:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">conserved_fitness</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_wavess.html">run_wavess</a></span><span class="op">(</span></span>
<span>  inf_pop_size <span class="op">=</span> <span class="va">pop</span>,</span>
<span>  samp_scheme <span class="op">=</span> <span class="va">samp</span>,</span>
<span>  founder_seqs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">founder_ref</span><span class="op">$</span><span class="va">founder</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>  conserved_sites <span class="op">=</span> <span class="va">founder_conserved_sites</span>,</span>
<span>  conserved_cost <span class="op">=</span> <span class="fl">0.99</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>If you look at the mean conserved fitness of active cells, you’ll see
that it’s not always 1 (although it may sometimes be 1 depending on the
simulation, since the output is stochastic).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">conserved_fitness</span><span class="op">$</span><span class="va">counts</span><span class="op">$</span><span class="va">mean_conserved_active</span></span>
<span><span class="co">#&gt; [1] 1.000000 1.000000 0.999505 1.000000</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="comparison-to-a-reference">Comparison to a reference<a class="anchor" aria-label="anchor" href="#comparison-to-a-reference"></a>
</h3>
<p>To simulate fitness compared to a reference, you must provide a “most
fit” reference sequence to compare each simulated sequence to. The
strength of this fitness can be altered using the
<code>replicative_cost</code> argument.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ref_fitness</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_wavess.html">run_wavess</a></span><span class="op">(</span></span>
<span>  inf_pop_size <span class="op">=</span> <span class="va">pop</span>,</span>
<span>  samp_scheme <span class="op">=</span> <span class="va">samp</span>,</span>
<span>  founder_seqs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">founder_ref</span><span class="op">$</span><span class="va">founder</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>  ref_seq <span class="op">=</span> <span class="va">founder_ref</span><span class="op">$</span><span class="va">ref</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Here, you can see that the mean replicative fitness is now less than
1:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ref_fitness</span><span class="op">$</span><span class="va">counts</span><span class="op">$</span><span class="va">mean_replicative_active</span></span>
<span><span class="co">#&gt; [1] 0.8520756 0.8493573 0.8460892 0.8430956</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="immune-fitness">Immune fitness<a class="anchor" aria-label="anchor" href="#immune-fitness"></a>
</h3>
<p>Active immunity can be modeled by defining B-cell epitopes locations
that can be recognized by the immune system. Nucleotide epitopes are
translated into amino acids before immune fitness costs are calculated.
An epitope is recognized only when it is present at least 100 times
across the population (<code>n_for_imm</code> argument). Once an epitope
is recognized, it undergoes affinity maturation for 90 days
(<code>days_full_potency</code>), at which time it reaches full potency.
Epitopes are also cross-reactive.</p>
<p>Cross-reactivity is sampled from a Beta distribution where alpha is 1
and beta is the square of the Hamming distance to the nearest recognized
epitope:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>    n_muts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>,</span>
<span>    quantile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span> <span class="op">/</span> <span class="fl">1000</span><span class="op">)</span>,</span>
<span>    val <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">dbeta</a></span><span class="op">(</span><span class="va">quantile</span>, <span class="fl">1</span>, <span class="va">x</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">quantile</span>, y <span class="op">=</span> <span class="va">val</span>, col <span class="op">=</span> <span class="va">n_muts</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Spectral"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Quantile"</span>, y <span class="op">=</span> <span class="st">"Value"</span>, col <span class="op">=</span> <span class="st">"Hamming\ndistance"</span><span class="op">)</span></span></code></pre></div>
<p><img src="run_wavess_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">immune_fitness</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_wavess.html">run_wavess</a></span><span class="op">(</span></span>
<span>  inf_pop_size <span class="op">=</span> <span class="va">pop</span>,</span>
<span>  samp_scheme <span class="op">=</span> <span class="va">samp</span>,</span>
<span>  founder_seqs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">founder_ref</span><span class="op">$</span><span class="va">founder</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>  epitope_locations <span class="op">=</span> <span class="va">epitope_locations</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The immune fitness here is now less than 1 after the immune system
kicks in:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">immune_fitness</span><span class="op">$</span><span class="va">counts</span><span class="op">$</span><span class="va">mean_immune_active</span></span>
<span><span class="co">#&gt; [1] 1.0000000 0.7033000 0.7284700 0.7380817</span></span></code></pre></div>
<p>Please note that the model is very sensitive to the maximum immune
fitness cost for a given epitope.</p>
</div>
</div>
<div class="section level2">
<h2 id="multiple-selective-pressures-at-once">Multiple selective pressures at once<a class="anchor" aria-label="anchor" href="#multiple-selective-pressures-at-once"></a>
</h2>
<p>You can also include multiple selective pressures at once by using
multiple of the arguments described above.</p>
</div>
<div class="section level2">
<h2 id="events-defined-by-rates">Events defined by rates<a class="anchor" aria-label="anchor" href="#events-defined-by-rates"></a>
</h2>
<p>There are several events that can occur during within-host evolution
that are defined by rates of occurrence (which are converted into
per-generation probabilities). These include the mutation rate
(<code>mut_rate</code>), the recombination rate
(<code>recomb_rate</code>), and various rates related to latent cell
dynamics (<code>act_to_lat</code>, <code>lat_to_act</code>,
<code>lat_prolif</code>, <code>lat_die</code>). Each of these can be
turned off by setting the respective rate value to 0. The simulation
starts with 0 latent cells, so setting <code>act_to_lat</code> will turn
latent cell dynamics off.</p>
</div>
<div class="section level2">
<h2 id="seed">Seed<a class="anchor" aria-label="anchor" href="#seed"></a>
</h2>
<p>A seed can be set for reproducibility using the <code>seed</code>
argument.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Zena Lapp, Narmada Sambaturu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
